// @super-protocol/sdk-js/build/utils/resourceLoaders/StorageProviderLoader.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StorageProviderLoader = void 0;
const fs_1 = __importDefault(require("fs"));
const stream_1 = __importDefault(require("stream"));
const dto_js_1 = require("@super-protocol/dto-js");
const ChunksDownloadDecorator_1 = require("../../providers/storage/ChunksDownloadDecorator");
const getStorageProvider_1 = __importDefault(require("../../providers/storage/getStorageProvider"));
const BaseResourceLoader_1 = require("./BaseResourceLoader");
class StorageProviderLoader extends BaseResourceLoader_1.BaseResourceLoader {
    constructor(config = {}) {
        var _a, _b, _c;
        super();
        this.progressListener = config.progressListener;
        this.chunkSize = config.chunkSize;
        this.concurrency = config.concurrency;
        this.startOffset = config.startOffset || StorageProviderLoader.DEFAULT_START_OFFSET;
        this.retry = {
            retryMaxCount: ((_a = config.retry) === null || _a === void 0 ? void 0 : _a.retryMaxCount) || StorageProviderLoader.DEFAULT_RETRY_COUNT,
            onRetry: (_b = config.retry) === null || _b === void 0 ? void 0 : _b.onRetry,
            retryWaitTimeFactory: (_c = config.retry) === null || _c === void 0 ? void 0 : _c.retryWaitTimeFactory,
        };
    }
    download(resource) {
        return __awaiter(this, void 0, void 0, function* () {
            const downloadStream = yield this.getFileStream(resource);
            return this.downloadToBuffer(downloadStream);
        });
    }
    downloadToFile(resource, downloadPath) {
        return __awaiter(this, void 0, void 0, function* () {
            const downloadStream = yield this.getFileStream(resource);
            yield stream_1.default.promises.pipeline(downloadStream, fs_1.default.createWriteStream(downloadPath));
        });
    }
    getFileStream(resource) {
        const storageProvider = (0, getStorageProvider_1.default)(resource);
        let storageProviderToInvoke;
        if (this.chunkSize && this.concurrency) {
            const chunkedDownloadWithRetry = (0, ChunksDownloadDecorator_1.createDownloadChunkMethodWithRetry)(ChunksDownloadDecorator_1.downloadChunkMethod, this.retry);
            storageProviderToInvoke = new ChunksDownloadDecorator_1.DownloadDecorator(storageProvider, chunkedDownloadWithRetry, {
                chunkSize: this.chunkSize,
                concurrency: this.concurrency,
                offset: this.startOffset,
            });
        }
        else {
            storageProviderToInvoke = storageProvider;
        }
        return storageProviderToInvoke.downloadFile(resource.filepath, {}, this.progressListener);
    }
}
exports.StorageProviderLoader = StorageProviderLoader;
StorageProviderLoader.DEFAULT_RETRY_COUNT = 3;
StorageProviderLoader.DEFAULT_START_OFFSET = 0;
StorageProviderLoader.type = dto_js_1.ResourceType.StorageProvider;
