// @super-protocol/sdk-js/build/crypto/nodejs/AES.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const dto_js_1 = require("@super-protocol/dto-js");
const NativeCrypto_1 = __importDefault(require("./NativeCrypto"));
class AES {
    static encrypt(content, encryption) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!encryption.key)
                throw Error('Encryption key is not provided');
            encryption.cipher = encryption.cipher || dto_js_1.Cipher.AES_256_GCM;
            const keyBuffer = Buffer.from(encryption.key, encryption.encoding);
            const encrypted = NativeCrypto_1.default.encrypt(keyBuffer, content, encryption.cipher);
            return {
                algo: dto_js_1.CryptoAlgorithm.AES,
                encoding: encryption.encoding,
                cipher: encryption.cipher,
                ciphertext: encrypted.ciphertext,
                iv: encrypted.iv,
                mac: encrypted.mac,
            };
        });
    }
    /**
     * Encrypts data stream
     * @param inputStream - path to file that will be encrypted
     * @param outputStream - place where it will be saved
     * @param algorithm - file encryption algorithm
     * @param key – key that will be used to encrypt data
     * @returns {Promise<Encryption>} - encryption info
     */
    static encryptStream(inputStream, outputStream, encryption) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!encryption.key)
                throw Error('Encryption key is not provided');
            encryption.cipher = encryption.cipher || dto_js_1.Cipher.AES_256_GCM;
            const keyBuffer = Buffer.from(encryption.key, encryption.encoding);
            const encrypted = yield NativeCrypto_1.default.encryptStream(keyBuffer, inputStream, outputStream, encryption.cipher);
            return {
                algo: encryption.algo,
                encoding: encryption.encoding,
                cipher: encryption.cipher,
                ciphertext: encrypted.ciphertext,
                iv: encrypted.iv,
                mac: encrypted.mac,
            };
        });
    }
    static decrypt(encryption) {
        if (!encryption.key)
            throw Error('Decryption key is not provided');
        const key = Buffer.from(encryption.key, encryption.encoding);
        const params = {
            iv: Buffer.from(encryption.iv, encryption.encoding),
        };
        if (encryption.mac) {
            params.mac = Buffer.from(encryption.mac, encryption.encoding);
        }
        return NativeCrypto_1.default.decrypt(key, encryption.ciphertext, encryption.cipher, params, encryption.encoding);
    }
    /**
     * Decrypts data stream
     * @param inputStream - stream with data to decrypt
     * @param outputStream - stream where the decrypted data will be written
     * @param encryption – encryption info
     */
    static decryptStream(inputStream, outputStream, encryption) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!encryption.key)
                throw Error('Decryption key is not provided');
            const key = Buffer.from(encryption.key, encryption.encoding);
            const params = {
                iv: Buffer.from(encryption.iv, encryption.encoding),
            };
            if (encryption.mac) {
                params.mac = Buffer.from(encryption.mac, encryption.encoding);
            }
            yield NativeCrypto_1.default.decryptStream(key, inputStream, outputStream, encryption.cipher, params);
        });
    }
}
exports.default = AES;
