// @super-protocol/sdk-js/build/crypto/nodejs/ECIES.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const dto_js_1 = require("@super-protocol/dto-js");
const crypto_1 = __importDefault(require("crypto"));
class ECIES {
    static encrypt(content, encryption) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!encryption.key)
                throw Error('Encryption key is not provided');
            const ecdh = crypto_1.default.createECDH('secp256k1');
            ecdh.generateKeys();
            const epk = ecdh.getPublicKey();
            const pk = ecdh.computeSecret(Buffer.from(encryption.key, encryption.encoding));
            const hash = crypto_1.default.createHash('sha512').update(pk).digest();
            const cipherKey = hash.slice(0, 32), macKey = hash.slice(32);
            const iv = crypto_1.default.randomBytes(16);
            const cipher = crypto_1.default.createCipheriv('aes-256-cbc', cipherKey, iv);
            let ct = cipher.update(content);
            ct = Buffer.concat([ct, cipher.final()]);
            const dataToMac = Buffer.concat([iv, epk, ct]);
            const mac = crypto_1.default.createHmac('sha256', macKey).update(dataToMac).digest();
            return {
                iv: iv.toString(encryption.encoding),
                ephemPublicKey: epk.toString(encryption.encoding),
                mac: mac.toString(encryption.encoding),
                encoding: encryption.encoding,
                algo: dto_js_1.CryptoAlgorithm.ECIES,
                ciphertext: ct.toString(encryption.encoding),
            };
        });
    }
    static decrypt(encryption) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!encryption.key)
                throw Error('Decryption key is not provided');
            const iv = Buffer.from(encryption.iv, encryption.encoding), epk = Buffer.from(encryption.ephemPublicKey, encryption.encoding), ct = Buffer.from(encryption.ciphertext, encryption.encoding), mac = Buffer.from(encryption.mac, encryption.encoding);
            const ecdh = crypto_1.default.createECDH('secp256k1');
            ecdh.setPrivateKey(Buffer.from(encryption.key, encryption.encoding));
            const pk = ecdh.computeSecret(epk);
            const hash = crypto_1.default.createHash('sha512').update(pk).digest();
            const cipherKey = hash.slice(0, 32), macKey = hash.slice(32);
            const m = crypto_1.default
                .createHmac('sha256', macKey)
                .update(Buffer.concat([iv, epk, ct]))
                .digest();
            if (m.compare(mac) !== 0 || mac.compare(m) !== 0) {
                throw new Error('Corrupted Ecies body: unmatched authentication code');
            }
            const decipher = crypto_1.default.createDecipheriv('aes-256-cbc', cipherKey, iv);
            const pt = decipher.update(ct);
            const result = Buffer.concat([pt, decipher.final()]);
            return result.toString('binary');
        });
    }
}
exports.default = ECIES;
