// @super-protocol/sdk-js/build/crypto/nodejs/NativeCrypto.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const crypto_1 = require("crypto");
const events_1 = require("events");
const dto_js_1 = require("@super-protocol/dto-js");
/**
 *
 */
class NativeCrypto {
    /**
     * Here would be better to check cipher type using
     * ```
     * getCipherInfo(cipher)!.keyLength
     * ```
     * but it doesn't work in browser
     */
    static getKeyLength(cipher) {
        if (/256\-xts/.test(cipher)) {
            return 64;
        }
        if (/256|128\-xts|chacha20/.test(cipher) && cipher !== 'aes-128-cbc-hmac-sha256') {
            return 32;
        }
        if (/192|des\-ede3|desx|des3$/.test(cipher) || cipher === 'id-smime-alg-cms3deswrap') {
            return 24;
        }
        if (/128|des\-ede/.test(cipher)) {
            return 16;
        }
        if (/64|des/.test(cipher)) {
            return 8;
        }
        if (/40/.test(cipher)) {
            return 5;
        }
        return 16;
    }
    /**
     * Here would be better to check cipher type using
     * ```
     * getCipherInfo(cipher)!.ivLength
     * ```
     * but it doesn't work in browser
     */
    static getIVLength(cipher) {
        if (this.isCCM(cipher) || this.isGCM(cipher) || this.isOCB(cipher)) {
            return 12;
        }
        if (/wrap\-pad/.test(cipher)) {
            return 4;
        }
        if (/wrap|cast|des|bf|blowfish|idea|rc2/.test(cipher)) {
            return 8;
        }
        return 16;
    }
    static createKey(cipher) {
        const length = this.getKeyLength(cipher);
        return (0, crypto_1.randomBytes)(length);
    }
    static createIV(cipher) {
        const length = this.getIVLength(cipher);
        return (0, crypto_1.randomBytes)(length);
    }
    static createCipher(cipher, key, iv) {
        if (this.isECB(cipher) || this.isRC4(cipher)) {
            return (0, crypto_1.createCipher)(cipher, key);
        }
        if (this.isCCM(cipher) || this.isOCB(cipher)) {
            const options = {
                authTagLength: 16,
            };
            return (0, crypto_1.createCipheriv)(cipher, key, iv, options);
        }
        return (0, crypto_1.createCipheriv)(cipher, key, iv);
    }
    static createDecipher(cipher, key, iv, mac) {
        if (iv) {
            const options = {};
            if (this.isCCM(cipher) || this.isOCB(cipher)) {
                options.authTagLength = 16;
            }
            const decipher = (0, crypto_1.createDecipheriv)(cipher, key, iv, options);
            if (mac) {
                decipher.setAuthTag(mac);
            }
            return decipher;
        }
        else {
            return (0, crypto_1.createDecipher)(cipher, key);
        }
    }
    static encrypt(key, content, cipherName, outputEncoding = dto_js_1.Encoding.base64, 
    // TODO: replace BufferEncoding with Encoding
    inputEncoding = 'binary') {
        const iv = this.createIV(cipherName);
        const result = {};
        const cipher = this.createCipher(cipherName, key, iv);
        result.ciphertext = cipher.update(content, inputEncoding, outputEncoding);
        result.ciphertext += cipher.final(outputEncoding);
        if (!this.isECB(cipherName) && !this.isRC4(cipherName)) {
            result.iv = iv.toString(outputEncoding);
        }
        if (this.isCCM(cipherName) || this.isGCM(cipherName) || this.isOCB(cipherName)) {
            result.mac = cipher.getAuthTag().toString(outputEncoding);
        }
        return result;
    }
    static encryptStream(key, inputStream, outputStream, cipherName, encoding = dto_js_1.Encoding.base64) {
        return __awaiter(this, void 0, void 0, function* () {
            const iv = this.createIV(cipherName);
            const result = {};
            const cipher = this.createCipher(cipherName, key, iv);
            inputStream.pipe(cipher).pipe(outputStream);
            yield (0, events_1.once)(outputStream, 'finish');
            result.iv = iv.toString(encoding);
            if (this.isCCM(cipherName) || this.isGCM(cipherName)) {
                result.mac = cipher.getAuthTag().toString(encoding);
            }
            return result;
        });
    }
    static decrypt(key, content, cipherName, params, inputEncoding = dto_js_1.Encoding.base64, 
    // TODO: replace BufferEncoding with Encoding
    outputEncoding = 'binary') {
        const decipher = this.createDecipher(cipherName, key, params === null || params === void 0 ? void 0 : params.iv, params === null || params === void 0 ? void 0 : params.mac);
        let decrypted = decipher.update(content, inputEncoding, outputEncoding);
        decrypted += decipher.final(outputEncoding);
        return decrypted;
    }
    static decryptStream(key, inputStream, outputStream, cipherName, params) {
        return __awaiter(this, void 0, void 0, function* () {
            const decipher = this.createDecipher(cipherName, key, params === null || params === void 0 ? void 0 : params.iv, params === null || params === void 0 ? void 0 : params.mac);
            inputStream.pipe(decipher).pipe(outputStream);
            yield (0, events_1.once)(outputStream, 'finish');
        });
    }
    static createHashFromBuffer(data, algorithm, encoding = dto_js_1.Encoding.base64) {
        const hash = (0, crypto_1.createHash)(algorithm);
        hash.update(data);
        return {
            algo: algorithm,
            encoding,
            hash: hash.digest(encoding),
        };
    }
    static createHashFromStream(inputStream, algorithm, encoding = dto_js_1.Encoding.base64) {
        const hash = (0, crypto_1.createHash)(algorithm);
        return new Promise((resolve) => {
            hash.on('finish', () => {
                resolve({
                    algo: algorithm,
                    encoding,
                    hash: hash.digest(encoding),
                });
            });
            inputStream.pipe(hash);
        });
    }
}
/**
 * Here would be better to check cipher type using
 * ```
 * getCipherInfo(cipher)!.mode === 'mode'
 * ```
 * but it doesn't work in browser
 */
NativeCrypto.isCCM = (cipher) => /ccm/i.test(cipher) || cipher === 'chacha20-poly1305';
NativeCrypto.isGCM = (cipher) => /gcm/i.test(cipher);
NativeCrypto.isOCB = (cipher) => /ocb/i.test(cipher);
NativeCrypto.isECB = (cipher) => /ecb/i.test(cipher) || cipher === 'des-ede' || cipher === 'des-ede3';
NativeCrypto.isRC4 = (cipher) => /^rc4/i.test(cipher);
exports.default = NativeCrypto;
