// @super-protocol/sdk-js/build/crypto/nodejs/RSA-Hybrid.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const dto_js_1 = require("@super-protocol/dto-js");
const crypto_1 = require("crypto");
const NativeCrypto_1 = __importDefault(require("./NativeCrypto"));
class RSAHybrid {
    static getKeyFingerprint(key) {
        const jsonKey = key.export({ format: 'jwk' });
        const buffer = Buffer.concat([
            Buffer.from('ssh-rsa'),
            Buffer.from(jsonKey.e || ''),
            Buffer.from(jsonKey.n || ''),
        ]);
        const hex = (0, crypto_1.createHash)('md5').update(buffer).digest('hex');
        return hex.match(/.{2}/g).join(':');
    }
    static publicEncrypt(rsaKey, aesKey) {
        const publicKey = (0, crypto_1.createPublicKey)({
            key: rsaKey,
            format: 'pem',
        });
        const encryptedKey = (0, crypto_1.publicEncrypt)({
            key: publicKey,
            padding: crypto_1.constants.RSA_PKCS1_OAEP_PADDING,
            oaepHash: 'sha256',
        }, Buffer.from(aesKey));
        const keyFingerprint = this.getKeyFingerprint(publicKey);
        return {
            [keyFingerprint]: encryptedKey.toString('base64'),
        };
    }
    static privateDecrypt(rsaKey, keys) {
        const privateKey = (0, crypto_1.createPrivateKey)({
            key: rsaKey,
            format: 'pem',
        });
        const keyFingerprint = this.getKeyFingerprint(privateKey);
        const encryptedKey = keys[keyFingerprint];
        const aesKey = (0, crypto_1.privateDecrypt)({
            key: privateKey,
            padding: crypto_1.constants.RSA_PKCS1_OAEP_PADDING,
            oaepHash: 'sha256',
        }, Buffer.from(encryptedKey, 'base64'));
        return aesKey;
    }
    static encrypt(content, encryption) {
        const aesKey = (0, crypto_1.randomBytes)(32);
        const keys = this.publicEncrypt(encryption.key, aesKey);
        const encrypted = NativeCrypto_1.default.encrypt(aesKey, content, dto_js_1.Cipher.AES_256_GCM, encryption.encoding);
        return Object.assign(Object.assign({}, encrypted), { encoding: encryption.encoding, algo: dto_js_1.CryptoAlgorithm.RSAHybrid, cipher: dto_js_1.Cipher.RSA_OAEP_AES_GCM, keys: JSON.stringify(keys) });
    }
    static encryptStream(inputStream, outputStream, encryption) {
        return __awaiter(this, void 0, void 0, function* () {
            const aesKey = (0, crypto_1.randomBytes)(32);
            const keys = this.publicEncrypt(encryption.key, aesKey);
            const encrypted = (yield NativeCrypto_1.default.encryptStream(aesKey, inputStream, outputStream, dto_js_1.Cipher.AES_256_GCM, encryption.encoding));
            return Object.assign(Object.assign({}, encrypted), { encoding: encryption.encoding, algo: dto_js_1.CryptoAlgorithm.RSAHybrid, cipher: dto_js_1.Cipher.RSA_OAEP_AES_GCM, keys: JSON.stringify(keys) });
        });
    }
    static decrypt(encryption) {
        const iv = Buffer.from(encryption.iv, encryption.encoding);
        const mac = Buffer.from(encryption.mac, encryption.encoding);
        const aesKey = this.privateDecrypt(encryption.key, JSON.parse(encryption.keys));
        const decrypted = NativeCrypto_1.default.decrypt(aesKey, encryption.ciphertext, dto_js_1.Cipher.AES_256_GCM, {
            iv,
            mac,
        }, encryption.encoding);
        return decrypted;
    }
    static decryptStream(inputStream, outputStream, encryption) {
        return __awaiter(this, void 0, void 0, function* () {
            const iv = Buffer.from(encryption.iv, encryption.encoding);
            const mac = Buffer.from(encryption.mac, encryption.encoding);
            const aesKey = this.privateDecrypt(encryption.key, JSON.parse(encryption.keys));
            yield NativeCrypto_1.default.decryptStream(aesKey, inputStream, outputStream, dto_js_1.Cipher.AES_256_GCM, {
                iv,
                mac,
            });
        });
    }
}
exports.default = RSAHybrid;
