// @super-protocol/sdk-js/build/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BLOCKCHAIN_BATCH_REQUEST_TIMEOUT = exports.BLOCKCHAIN_CALL_RETRY_INTERVAL = exports.BLOCKCHAIN_CALL_RETRY_ATTEMPTS = exports.POLYGON_MATIC_EVENT_PATH = exports.BLOCK_SIZE_TO_FETCH_TRANSACTION = exports.ONE_DAY = exports.txIntervalMs = exports.txConcurrency = exports.defaultGasLimitMultiplier = exports.defaultGasPriceMultiplier = exports.defaultGasPrice = exports.defaultGasLimit = exports.defaultBlockchainUrl = void 0;
exports.defaultBlockchainUrl = 'http://127.0.0.1:8545';
exports.defaultGasLimit = BigInt(7000000);
exports.defaultGasPrice = BigInt(3000000000);
exports.defaultGasPriceMultiplier = 1;
exports.defaultGasLimitMultiplier = 1.2;
exports.txConcurrency = 30;
exports.txIntervalMs = 34;
exports.ONE_DAY = 24 * 60 * 60;
exports.BLOCK_SIZE_TO_FETCH_TRANSACTION = 500;
exports.POLYGON_MATIC_EVENT_PATH = '0x0000000000000000000000000000000000001010';
exports.BLOCKCHAIN_CALL_RETRY_ATTEMPTS = 10;
exports.BLOCKCHAIN_CALL_RETRY_INTERVAL = 3000;
exports.BLOCKCHAIN_BATCH_REQUEST_TIMEOUT = 10000;
