// @super-protocol/sdk-js/build/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QuoteValidator = exports.TeeSgxParser = exports.Web3TransactionRevertedByEvmError = exports.helpers = exports.getStorageProvider = exports.ContentWriterType = exports.StorageContentWriter = exports.CacheEvents = exports.StorageAdapter = exports.StorageKeyValueAdapter = exports.StorageMetadataReader = exports.StorjAdapter = exports.ChunkedStorageProvider = exports.TCB = exports.Offer = exports.TeeOffer = exports.Provider = exports.Order = exports.Deposits = exports.Marks = exports.Superpro = exports.SuperproToken = exports.ProviderRegistry = exports.TeeOffers = exports.Offers = exports.ActiveOffers = exports.ActiveOrders = exports.Orders = exports.Consensus = exports.TIIGenerator = exports.Crypto = exports.BlockchainEventsListener = void 0;
const BlockchainConnector_1 = __importDefault(require("./connectors/BlockchainConnector"));
exports.default = BlockchainConnector_1.default;
__exportStar(require("./connectors/BlockchainConnector"), exports);
var BlockchainEventsListener_1 = require("./connectors/BlockchainEventsListener");
Object.defineProperty(exports, "BlockchainEventsListener", { enumerable: true, get: function () { return __importDefault(BlockchainEventsListener_1).default; } });
var crypto_1 = require("./crypto");
Object.defineProperty(exports, "Crypto", { enumerable: true, get: function () { return __importDefault(crypto_1).default; } });
var TIIGenerator_1 = require("./TIIGenerator");
Object.defineProperty(exports, "TIIGenerator", { enumerable: true, get: function () { return __importDefault(TIIGenerator_1).default; } });
__exportStar(require("./TIIGenerator"), exports);
var Consensus_1 = require("./staticModels/Consensus");
Object.defineProperty(exports, "Consensus", { enumerable: true, get: function () { return __importDefault(Consensus_1).default; } });
__exportStar(require("./staticModels/Consensus"), exports);
var Orders_1 = require("./staticModels/Orders");
Object.defineProperty(exports, "Orders", { enumerable: true, get: function () { return __importDefault(Orders_1).default; } });
__exportStar(require("./staticModels/Orders"), exports);
var ActiveOrders_1 = require("./staticModels/ActiveOrders");
Object.defineProperty(exports, "ActiveOrders", { enumerable: true, get: function () { return __importDefault(ActiveOrders_1).default; } });
__exportStar(require("./staticModels/ActiveOrders"), exports);
var ActiveOffers_1 = require("./staticModels/ActiveOffers");
Object.defineProperty(exports, "ActiveOffers", { enumerable: true, get: function () { return __importDefault(ActiveOffers_1).default; } });
__exportStar(require("./staticModels/ActiveOffers"), exports);
var Offers_1 = require("./staticModels/Offers");
Object.defineProperty(exports, "Offers", { enumerable: true, get: function () { return __importDefault(Offers_1).default; } });
__exportStar(require("./staticModels/Offers"), exports);
var TeeOffers_1 = require("./staticModels/TeeOffers");
Object.defineProperty(exports, "TeeOffers", { enumerable: true, get: function () { return __importDefault(TeeOffers_1).default; } });
__exportStar(require("./staticModels/TeeOffers"), exports);
var ProviderRegistry_1 = require("./staticModels/ProviderRegistry");
Object.defineProperty(exports, "ProviderRegistry", { enumerable: true, get: function () { return __importDefault(ProviderRegistry_1).default; } });
__exportStar(require("./staticModels/ProviderRegistry"), exports);
var SuperproToken_1 = require("./staticModels/SuperproToken");
Object.defineProperty(exports, "SuperproToken", { enumerable: true, get: function () { return __importDefault(SuperproToken_1).default; } });
__exportStar(require("./staticModels/SuperproToken"), exports);
var Superpro_1 = require("./staticModels/Superpro");
Object.defineProperty(exports, "Superpro", { enumerable: true, get: function () { return __importDefault(Superpro_1).default; } });
__exportStar(require("./staticModels/Superpro"), exports);
var Marks_1 = require("./staticModels/Marks");
Object.defineProperty(exports, "Marks", { enumerable: true, get: function () { return __importDefault(Marks_1).default; } });
__exportStar(require("./staticModels/Marks"), exports);
var Deposits_1 = require("./staticModels/Deposits");
Object.defineProperty(exports, "Deposits", { enumerable: true, get: function () { return __importDefault(Deposits_1).default; } });
__exportStar(require("./staticModels/Deposits"), exports);
var Order_1 = require("./models/Order");
Object.defineProperty(exports, "Order", { enumerable: true, get: function () { return __importDefault(Order_1).default; } });
__exportStar(require("./models/Order"), exports);
var Provider_1 = require("./models/Provider");
Object.defineProperty(exports, "Provider", { enumerable: true, get: function () { return __importDefault(Provider_1).default; } });
__exportStar(require("./models/Provider"), exports);
var TeeOffer_1 = require("./models/TeeOffer");
Object.defineProperty(exports, "TeeOffer", { enumerable: true, get: function () { return __importDefault(TeeOffer_1).default; } });
__exportStar(require("./models/TeeOffer"), exports);
var Offer_1 = require("./models/Offer");
Object.defineProperty(exports, "Offer", { enumerable: true, get: function () { return __importDefault(Offer_1).default; } });
__exportStar(require("./models/Offer"), exports);
var TCB_1 = require("./models/TCB");
Object.defineProperty(exports, "TCB", { enumerable: true, get: function () { return __importDefault(TCB_1).default; } });
__exportStar(require("./models/TCB"), exports);
__exportStar(require("./types/Offer"), exports);
__exportStar(require("./types/Marks"), exports);
__exportStar(require("./types/Order"), exports);
__exportStar(require("./types/TeeOfferInfo"), exports);
__exportStar(require("./types/HardwareInfo"), exports);
__exportStar(require("./types/SlotInfo"), exports);
__exportStar(require("./types/SlotUsage"), exports);
__exportStar(require("./types/OptionInfo"), exports);
__exportStar(require("./types/TeeOfferInfo"), exports);
__exportStar(require("./types/TeeOfferOption"), exports);
__exportStar(require("./types/TeeOfferSlot"), exports);
__exportStar(require("./types/ValueOfferSlot"), exports);
__exportStar(require("./types/Provider"), exports);
__exportStar(require("./types/Superpro"), exports);
__exportStar(require("./types/Consensus"), exports);
__exportStar(require("./types/Origins"), exports);
__exportStar(require("./types/Web3"), exports);
exports.ChunkedStorageProvider = __importStar(require("./providers/storage/ChunksDownloadDecorator"));
var StorjAdapter_1 = require("./providers/storage/StorjAdapter");
Object.defineProperty(exports, "StorjAdapter", { enumerable: true, get: function () { return __importDefault(StorjAdapter_1).default; } });
var StorageMetadataReader_1 = require("./providers/storage/StorageMetadataReader");
Object.defineProperty(exports, "StorageMetadataReader", { enumerable: true, get: function () { return __importDefault(StorageMetadataReader_1).default; } });
var StorageKeyValueAdapter_1 = require("./providers/storage/StorageKeyValueAdapter");
Object.defineProperty(exports, "StorageKeyValueAdapter", { enumerable: true, get: function () { return __importDefault(StorageKeyValueAdapter_1).default; } });
var StorageAdapter_1 = require("./providers/storage/StorageAdapter");
Object.defineProperty(exports, "StorageAdapter", { enumerable: true, get: function () { return __importDefault(StorageAdapter_1).default; } });
Object.defineProperty(exports, "CacheEvents", { enumerable: true, get: function () { return StorageAdapter_1.CacheEvents; } });
var StorageContentWriter_1 = require("./providers/storage/StorageContentWriter");
Object.defineProperty(exports, "StorageContentWriter", { enumerable: true, get: function () { return __importDefault(StorageContentWriter_1).default; } });
Object.defineProperty(exports, "ContentWriterType", { enumerable: true, get: function () { return StorageContentWriter_1.ContentWriterType; } });
var getStorageProvider_1 = require("./providers/storage/getStorageProvider");
Object.defineProperty(exports, "getStorageProvider", { enumerable: true, get: function () { return __importDefault(getStorageProvider_1).default; } });
exports.helpers = __importStar(require("./utils/helpers"));
var TxManager_1 = require("./utils/TxManager");
Object.defineProperty(exports, "Web3TransactionRevertedByEvmError", { enumerable: true, get: function () { return TxManager_1.Web3TransactionRevertedByEvmError; } });
require("./polyfills");
var QuoteParser_1 = require("./tee/QuoteParser");
Object.defineProperty(exports, "TeeSgxParser", { enumerable: true, get: function () { return QuoteParser_1.TeeSgxParser; } });
var QuoteValidator_1 = require("./tee/QuoteValidator");
Object.defineProperty(exports, "QuoteValidator", { enumerable: true, get: function () { return QuoteValidator_1.QuoteValidator; } });
