// @super-protocol/sdk-js/build/proto/Compression.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Compression = exports.compression_TYPEToJSON = exports.compression_TYPEFromJSON = exports.Compression_TYPE = exports.protobufPackage = void 0;
/* eslint-disable */
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "";
var Compression_TYPE;
(function (Compression_TYPE) {
    Compression_TYPE[Compression_TYPE["Uncompressed"] = 0] = "Uncompressed";
    Compression_TYPE[Compression_TYPE["GZIP"] = 1] = "GZIP";
})(Compression_TYPE = exports.Compression_TYPE || (exports.Compression_TYPE = {}));
function compression_TYPEFromJSON(object) {
    switch (object) {
        case 0:
        case "Uncompressed":
            return Compression_TYPE.Uncompressed;
        case 1:
        case "GZIP":
            return Compression_TYPE.GZIP;
        default:
            throw new globalThis.Error("Unrecognized enum value " + object + " for enum Compression_TYPE");
    }
}
exports.compression_TYPEFromJSON = compression_TYPEFromJSON;
function compression_TYPEToJSON(object) {
    switch (object) {
        case Compression_TYPE.Uncompressed:
            return "Uncompressed";
        case Compression_TYPE.GZIP:
            return "GZIP";
        default:
            throw new globalThis.Error("Unrecognized enum value " + object + " for enum Compression_TYPE");
    }
}
exports.compression_TYPEToJSON = compression_TYPEToJSON;
function createBaseCompression() {
    return { type: 0, data: new Uint8Array() };
}
exports.Compression = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.type !== 0) {
            writer.uint32(8).int32(message.type);
        }
        if (message.data.length !== 0) {
            writer.uint32(18).bytes(message.data);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCompression();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.type = reader.int32();
                    break;
                case 2:
                    message.data = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            type: isSet(object.type) ? compression_TYPEFromJSON(object.type) : 0,
            data: isSet(object.data) ? bytesFromBase64(object.data) : new Uint8Array(),
        };
    },
    toJSON(message) {
        const obj = {};
        message.type !== undefined && (obj.type = compression_TYPEToJSON(message.type));
        message.data !== undefined &&
            (obj.data = base64FromBytes(message.data !== undefined ? message.data : new Uint8Array()));
        return obj;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseCompression();
        message.type = (_a = object.type) !== null && _a !== void 0 ? _a : 0;
        message.data = (_b = object.data) !== null && _b !== void 0 ? _b : new Uint8Array();
        return message;
    },
};
var globalThis = (() => {
    if (typeof globalThis !== "undefined")
        return globalThis;
    if (typeof self !== "undefined")
        return self;
    if (typeof window !== "undefined")
        return window;
    if (typeof global !== "undefined")
        return global;
    throw "Unable to locate global object";
})();
function bytesFromBase64(b64) {
    if (globalThis.Buffer) {
        return Uint8Array.from(globalThis.Buffer.from(b64, "base64"));
    }
    else {
        const bin = globalThis.atob(b64);
        const arr = new Uint8Array(bin.length);
        for (let i = 0; i < bin.length; ++i) {
            arr[i] = bin.charCodeAt(i);
        }
        return arr;
    }
}
function base64FromBytes(arr) {
    if (globalThis.Buffer) {
        return globalThis.Buffer.from(arr).toString("base64");
    }
    else {
        const bin = [];
        arr.forEach((byte) => {
            bin.push(String.fromCharCode(byte));
        });
        return globalThis.btoa(bin.join(""));
    }
}
function isSet(value) {
    return value !== null && value !== undefined;
}
