// @super-protocol/sdk-js/build/proto/TRI.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TRI = exports.Hash = exports.Encryption = exports.protobufPackage = void 0;
/* eslint-disable */
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "";
function createBaseEncryption() {
    return {
        algo: "",
        key: undefined,
        cipher: undefined,
        ciphertext: undefined,
        iv: undefined,
        mac: undefined,
        encoding: "",
    };
}
exports.Encryption = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.algo !== "") {
            writer.uint32(10).string(message.algo);
        }
        if (message.key !== undefined) {
            writer.uint32(18).bytes(message.key);
        }
        if (message.cipher !== undefined) {
            writer.uint32(26).string(message.cipher);
        }
        if (message.ciphertext !== undefined) {
            writer.uint32(34).bytes(message.ciphertext);
        }
        if (message.iv !== undefined) {
            writer.uint32(50).bytes(message.iv);
        }
        if (message.mac !== undefined) {
            writer.uint32(58).bytes(message.mac);
        }
        if (message.encoding !== "") {
            writer.uint32(66).string(message.encoding);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseEncryption();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.algo = reader.string();
                    break;
                case 2:
                    message.key = reader.bytes();
                    break;
                case 3:
                    message.cipher = reader.string();
                    break;
                case 4:
                    message.ciphertext = reader.bytes();
                    break;
                case 6:
                    message.iv = reader.bytes();
                    break;
                case 7:
                    message.mac = reader.bytes();
                    break;
                case 8:
                    message.encoding = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            algo: isSet(object.algo) ? String(object.algo) : "",
            key: isSet(object.key) ? bytesFromBase64(object.key) : undefined,
            cipher: isSet(object.cipher) ? String(object.cipher) : undefined,
            ciphertext: isSet(object.ciphertext) ? bytesFromBase64(object.ciphertext) : undefined,
            iv: isSet(object.iv) ? bytesFromBase64(object.iv) : undefined,
            mac: isSet(object.mac) ? bytesFromBase64(object.mac) : undefined,
            encoding: isSet(object.encoding) ? String(object.encoding) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.algo !== undefined && (obj.algo = message.algo);
        message.key !== undefined && (obj.key = message.key !== undefined ? base64FromBytes(message.key) : undefined);
        message.cipher !== undefined && (obj.cipher = message.cipher);
        message.ciphertext !== undefined &&
            (obj.ciphertext = message.ciphertext !== undefined ? base64FromBytes(message.ciphertext) : undefined);
        message.iv !== undefined && (obj.iv = message.iv !== undefined ? base64FromBytes(message.iv) : undefined);
        message.mac !== undefined && (obj.mac = message.mac !== undefined ? base64FromBytes(message.mac) : undefined);
        message.encoding !== undefined && (obj.encoding = message.encoding);
        return obj;
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g;
        const message = createBaseEncryption();
        message.algo = (_a = object.algo) !== null && _a !== void 0 ? _a : "";
        message.key = (_b = object.key) !== null && _b !== void 0 ? _b : undefined;
        message.cipher = (_c = object.cipher) !== null && _c !== void 0 ? _c : undefined;
        message.ciphertext = (_d = object.ciphertext) !== null && _d !== void 0 ? _d : undefined;
        message.iv = (_e = object.iv) !== null && _e !== void 0 ? _e : undefined;
        message.mac = (_f = object.mac) !== null && _f !== void 0 ? _f : undefined;
        message.encoding = (_g = object.encoding) !== null && _g !== void 0 ? _g : "";
        return message;
    },
};
function createBaseHash() {
    return { algo: "", hash: new Uint8Array() };
}
exports.Hash = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.algo !== "") {
            writer.uint32(10).string(message.algo);
        }
        if (message.hash.length !== 0) {
            writer.uint32(18).bytes(message.hash);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseHash();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.algo = reader.string();
                    break;
                case 2:
                    message.hash = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            algo: isSet(object.algo) ? String(object.algo) : "",
            hash: isSet(object.hash) ? bytesFromBase64(object.hash) : new Uint8Array(),
        };
    },
    toJSON(message) {
        const obj = {};
        message.algo !== undefined && (obj.algo = message.algo);
        message.hash !== undefined &&
            (obj.hash = base64FromBytes(message.hash !== undefined ? message.hash : new Uint8Array()));
        return obj;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseHash();
        message.algo = (_a = object.algo) !== null && _a !== void 0 ? _a : "";
        message.hash = (_b = object.hash) !== null && _b !== void 0 ? _b : new Uint8Array();
        return message;
    },
};
function createBaseTRI() {
    return { solutionHashes: [], mrenclave: new Uint8Array(), args: "", encryption: undefined };
}
exports.TRI = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.solutionHashes) {
            exports.Hash.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (message.mrenclave.length !== 0) {
            writer.uint32(18).bytes(message.mrenclave);
        }
        if (message.args !== "") {
            writer.uint32(26).string(message.args);
        }
        if (message.encryption !== undefined) {
            exports.Encryption.encode(message.encryption, writer.uint32(34).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTRI();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.solutionHashes.push(exports.Hash.decode(reader, reader.uint32()));
                    break;
                case 2:
                    message.mrenclave = reader.bytes();
                    break;
                case 3:
                    message.args = reader.string();
                    break;
                case 4:
                    message.encryption = exports.Encryption.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            solutionHashes: Array.isArray(object === null || object === void 0 ? void 0 : object.solutionHashes)
                ? object.solutionHashes.map((e) => exports.Hash.fromJSON(e))
                : [],
            mrenclave: isSet(object.mrenclave) ? bytesFromBase64(object.mrenclave) : new Uint8Array(),
            args: isSet(object.args) ? String(object.args) : "",
            encryption: isSet(object.encryption) ? exports.Encryption.fromJSON(object.encryption) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.solutionHashes) {
            obj.solutionHashes = message.solutionHashes.map((e) => (e ? exports.Hash.toJSON(e) : undefined));
        }
        else {
            obj.solutionHashes = [];
        }
        message.mrenclave !== undefined &&
            (obj.mrenclave = base64FromBytes(message.mrenclave !== undefined ? message.mrenclave : new Uint8Array()));
        message.args !== undefined && (obj.args = message.args);
        message.encryption !== undefined &&
            (obj.encryption = message.encryption ? exports.Encryption.toJSON(message.encryption) : undefined);
        return obj;
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseTRI();
        message.solutionHashes = ((_a = object.solutionHashes) === null || _a === void 0 ? void 0 : _a.map((e) => exports.Hash.fromPartial(e))) || [];
        message.mrenclave = (_b = object.mrenclave) !== null && _b !== void 0 ? _b : new Uint8Array();
        message.args = (_c = object.args) !== null && _c !== void 0 ? _c : "";
        message.encryption =
            object.encryption !== undefined && object.encryption !== null
                ? exports.Encryption.fromPartial(object.encryption)
                : undefined;
        return message;
    },
};
var globalThis = (() => {
    if (typeof globalThis !== "undefined")
        return globalThis;
    if (typeof self !== "undefined")
        return self;
    if (typeof window !== "undefined")
        return window;
    if (typeof global !== "undefined")
        return global;
    throw "Unable to locate global object";
})();
function bytesFromBase64(b64) {
    if (globalThis.Buffer) {
        return Uint8Array.from(globalThis.Buffer.from(b64, "base64"));
    }
    else {
        const bin = globalThis.atob(b64);
        const arr = new Uint8Array(bin.length);
        for (let i = 0; i < bin.length; ++i) {
            arr[i] = bin.charCodeAt(i);
        }
        return arr;
    }
}
function base64FromBytes(arr) {
    if (globalThis.Buffer) {
        return globalThis.Buffer.from(arr).toString("base64");
    }
    else {
        const bin = [];
        arr.forEach((byte) => {
            bin.push(String.fromCharCode(byte));
        });
        return globalThis.btoa(bin.join(""));
    }
}
function isSet(value) {
    return value !== null && value !== undefined;
}
