// @super-protocol/sdk-js/build/connectors/BaseConnector.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseConnector = void 0;
const logger_1 = __importDefault(require("../logger"));
class BaseConnector {
    constructor() {
        this.initialized = false;
        this.logger = logger_1.default.child({ className: this.constructor['name'] });
    }
    isInitialized() {
        return this.initialized;
    }
    checkIfInitialized() {
        if (!this.initialized)
            throw new Error(`${this.constructor['name']} is not initialized, needs to run '${this.constructor['name']}.initialize(CONFIG)' first`);
    }
    /**
     *
     * @returns initialized contract
     */
    getContract() {
        this.checkIfInitialized();
        return this.contract;
    }
    /**
     * Function for connecting to blockchain
     * Used to setting up settings for blockchain connector
     * Needs to run this function before using blockchain connector
     */
    initialize(_config) {
        return __awaiter(this, void 0, void 0, function* () { });
    }
    shutdown() {
        if (this.initialized) {
            this.initialized = false;
            this.logger.trace(`${this.constructor['name']} was shutdown`);
        }
    }
}
exports.BaseConnector = BaseConnector;
