// @super-protocol/sdk-js/build/connectors/BlockchainEventsListener.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const BaseConnector_1 = require("./BaseConnector");
const web3_1 = __importStar(require("web3"));
const abi_1 = require("../contracts/abi");
// TODO: remove this dependencies
const store_1 = __importDefault(require("../store"));
const Superpro_1 = __importDefault(require("../staticModels/Superpro"));
const SuperproToken_1 = __importDefault(require("../staticModels/SuperproToken"));
class BlockchainEventsListener extends BaseConnector_1.BaseConnector {
    constructor() {
        super();
    }
    static getInstance() {
        if (!BlockchainEventsListener.instance) {
            BlockchainEventsListener.instance = new BlockchainEventsListener();
        }
        return BlockchainEventsListener.instance;
    }
    getProvider() {
        var _a;
        return (_a = store_1.default.web3Wss) === null || _a === void 0 ? void 0 : _a.provider;
    }
    /**
     * Function for connecting to blockchain using web socket
     * Needs to run this function before using events
     */
    initialize(config) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.trace(config, 'Initializing');
            const reconnectOptions = Object.assign({
                autoReconnect: true,
                delay: 20000,
                maxAttempts: 5000,
            }, config.reconnect);
            this.logger.info(`Initializing events listener with reconnect options: ${JSON.stringify(reconnectOptions)}`);
            const provider = new web3_1.WebSocketProvider(config.blockchainUrl, {} /* ClientOptions */, reconnectOptions);
            store_1.default.web3Wss = new web3_1.default(provider);
            const web3Context = new web3_1.Web3Context({
                provider: store_1.default.web3Wss.currentProvider,
                config: { contractDataInputFill: 'data' },
            });
            this.contract = new store_1.default.web3Wss.eth.Contract(abi_1.abi, config.contractAddress, web3Context);
            Superpro_1.default.address = config.contractAddress;
            SuperproToken_1.default.addressWss = yield Superpro_1.default.getTokenAddress(this.contract);
            this.initialized = true;
            this.logger.trace('Initialized');
        });
    }
    shutdown() {
        var _a, _b;
        super.shutdown();
        (_b = (_a = store_1.default.web3Wss) === null || _a === void 0 ? void 0 : _a.provider) === null || _b === void 0 ? void 0 : _b.disconnect();
        store_1.default.web3Wss = undefined;
    }
}
exports.default = BlockchainEventsListener;
