// @super-protocol/tee-lib/dist/tlb/tee-sgx-parserv3.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TeeSgxParserV3 = void 0;
const buffer_1 = require("buffer/");
const errors_1 = require("./errors");
/**
 * Sgx quote parser (quote 3 version).
 *
 * https://download.01.org/intel-sgx/dcap-1.0.1/docs/Intel_SGX_ECDSA_QuoteGenReference_DCAP_API_Linux_1.0.1.pdf
 */
class TeeSgxParserV3 {
    getDataAndAdvance(blob, size) {
        const buf = buffer_1.Buffer.from(blob.data.subarray(0, size));
        blob.data = buffer_1.Buffer.from(blob.data.subarray(size));
        return buf;
    }
    parseQuote(data) {
        const { quoteHeaderSize, reportSize, userDataOffset, userDataSize, ecdsaP256SignatureSize, ecdsaP256PublicKeySize, } = TeeSgxParserV3;
        if (data.length < quoteHeaderSize + reportSize) {
            throw new errors_1.TeeSgxParserError('data has invalid length');
        }
        const quoteRemainder = { data: buffer_1.Buffer.from(data) };
        const quoteHeader = this.getDataAndAdvance(quoteRemainder, quoteHeaderSize);
        const report = this.getDataAndAdvance(quoteRemainder, reportSize);
        const version = quoteHeader.readUInt16LE(0);
        if (version != 3) {
            throw new errors_1.TeeSgxParserError('quote header has invalid or unsupported version');
        }
        const attestationKeyType = quoteHeader.readUInt16LE(2);
        if (attestationKeyType > 3) {
            throw new errors_1.TeeSgxParserError('quote header has invalid or unsupported attestation key type');
        }
        const userData = quoteHeader.slice(userDataOffset, userDataOffset + userDataSize);
        const quoteSignatureDateLen = quoteRemainder.data.readUInt32LE(0);
        quoteRemainder.data = buffer_1.Buffer.from(quoteRemainder.data.subarray(4));
        if (quoteSignatureDateLen != quoteRemainder.data.length) {
            throw new errors_1.TeeSgxParserError(`quoteSignatureDateLen has invalid length: ${quoteRemainder.data.length} instead of ${quoteSignatureDateLen} expected`);
        }
        const rawQuoteSignatureDataRemainder = {
            data: this.getDataAndAdvance(quoteRemainder, quoteSignatureDateLen),
        };
        const isvEnclaveReportSignature = this.getDataAndAdvance(rawQuoteSignatureDataRemainder, ecdsaP256SignatureSize);
        const ecdsaAttestationKey = this.getDataAndAdvance(rawQuoteSignatureDataRemainder, ecdsaP256PublicKeySize);
        const qeReport = this.getDataAndAdvance(rawQuoteSignatureDataRemainder, reportSize);
        const qeReportSignature = this.getDataAndAdvance(rawQuoteSignatureDataRemainder, ecdsaP256SignatureSize);
        const qeAuthenticationDataSize = rawQuoteSignatureDataRemainder.data.readUInt16LE(0);
        rawQuoteSignatureDataRemainder.data = buffer_1.Buffer.from(rawQuoteSignatureDataRemainder.data.subarray(2));
        if (rawQuoteSignatureDataRemainder.data.length < qeAuthenticationDataSize) {
            throw new errors_1.TeeSgxParserError(`qeAuthenticationDataSize has invalid length: ${rawQuoteSignatureDataRemainder.data.length} instead of ${qeAuthenticationDataSize} expected`);
        }
        const qeAuthenticationData = this.getDataAndAdvance(rawQuoteSignatureDataRemainder, qeAuthenticationDataSize);
        const qeCertificationDataType = rawQuoteSignatureDataRemainder.data.readUInt16LE(0);
        if (qeCertificationDataType < 1 || qeCertificationDataType > 7) {
            throw new errors_1.TeeSgxParserError(`certificationDataType has invalid value: ${qeCertificationDataType}`);
        }
        const certificationDataSize = rawQuoteSignatureDataRemainder.data.readUInt32LE(2);
        const qeCertificationData = rawQuoteSignatureDataRemainder.data.subarray(2 + 4);
        if (certificationDataSize != qeCertificationData.length) {
            throw new errors_1.TeeSgxParserError(`certificationDataSize has invalid length: $PqeCertificationData.length} instead of ${certificationDataSize} expected`);
        }
        return {
            rawHeader: quoteHeader,
            header: {
                version,
                attestationKeyType,
                userData,
            },
            report,
            isvEnclaveReportSignature,
            ecdsaAttestationKey,
            qeReport,
            qeReportSignature,
            qeAuthenticationData,
            qeCertificationDataType,
            qeCertificationData,
        };
    }
    parseReport(data) {
        const { reportSize, cpuSvnSize, reportMrEnclaveOffset, reportMrEnclaveSize, reportMrSignerOffset, reportMrSignerSize, reportDataOffset, reportUserDataSize, reportUserDataSHA256Size, } = TeeSgxParserV3;
        if (data.length < reportSize) {
            throw new errors_1.TeeSgxParserError('data has invalid length');
        }
        const report = buffer_1.Buffer.from(data);
        const cpuSvn = report.slice(0, cpuSvnSize).toString('hex');
        const mrEnclave = report.slice(reportMrEnclaveOffset, reportMrEnclaveOffset + reportMrEnclaveSize);
        const mrSigner = report.slice(reportMrSignerOffset, reportMrSignerOffset + reportMrSignerSize);
        const userData = report.slice(reportDataOffset, reportDataOffset + reportUserDataSize);
        const dataHash = report.slice(reportDataOffset, reportDataOffset + reportUserDataSHA256Size);
        return {
            cpuSvn,
            mrEnclave,
            mrSigner,
            userData,
            dataHash,
        };
    }
}
exports.TeeSgxParserV3 = TeeSgxParserV3;
TeeSgxParserV3.quoteHeaderSize = 48;
TeeSgxParserV3.reportSize = 384;
TeeSgxParserV3.userDataOffset = 28;
TeeSgxParserV3.userDataSize = 20;
TeeSgxParserV3.cpuSvnSize = 16;
TeeSgxParserV3.reportMrEnclaveOffset = 64;
TeeSgxParserV3.reportMrEnclaveSize = 32;
TeeSgxParserV3.reportMrSignerOffset = TeeSgxParserV3.reportMrEnclaveOffset + TeeSgxParserV3.reportMrEnclaveSize + /* reserved */ 32;
TeeSgxParserV3.reportMrSignerSize = 32;
TeeSgxParserV3.reportDataOffset = TeeSgxParserV3.reportMrSignerOffset +
    TeeSgxParserV3.reportMrSignerSize +
    /* reserved */ 96 +
    /* ISVProdID */ 2 +
    /* ISVSVN */ 2 +
    /* reserved */ 60;
TeeSgxParserV3.reportUserDataSize = 64;
TeeSgxParserV3.reportUserDataSHA256Size = 32; /* 64 in report, but we need 32 only for sha256 hash */
TeeSgxParserV3.ecdsaP256SignatureSize = 64;
TeeSgxParserV3.ecdsaP256PublicKeySize = 64;
