// @super-protocol/tee-lib/dist/tlb/tlblock-serializerv1.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TLBlockSerializerV1 = void 0;
const gzip = __importStar(require("gzip-js"));
const msgpack_1 = require("@msgpack/msgpack");
const errors_1 = require("./errors");
const SERIALIZER_VERSION = 1;
class TLBlockSerializerV1 {
    unserializeTlb(blob) {
        const tlb = (0, msgpack_1.decode)(blob, {
            maxBinLength: 8192,
            maxArrayLength: 256,
            maxExtLength: 256,
            maxMapLength: 128,
            maxStrLength: 128,
        });
        if (tlb.v !== SERIALIZER_VERSION) {
            throw new errors_1.TLBlockSerializerError('tlb has invalid or unsupported version');
        }
        const data = (0, msgpack_1.decode)(tlb.data);
        return {
            quote: data.quote.format === 'gzip' ? new Uint8Array(gzip.unzip(tlb.quote)) : tlb.quote,
            data,
            dataBlob: tlb.data,
        };
    }
    serializeTlb(tlb, tlbMetadata) {
        const processedQuote = tlbMetadata.quote.format === 'gzip' ? gzip.zip(tlb.quote, { level: 9 }) : tlb.quote;
        return (0, msgpack_1.encode)({
            v: SERIALIZER_VERSION,
            quote: new Uint8Array(processedQuote),
            data: new Uint8Array(tlb.data),
        });
    }
    serializeMetadata(tlbMetadata) {
        return (0, msgpack_1.encode)(tlbMetadata, { sortKeys: true });
    }
    serializeAnyData(data) {
        const quote = {
            v: SERIALIZER_VERSION,
            quote: data,
        };
        return (0, msgpack_1.encode)(quote, { sortKeys: true });
    }
}
exports.TLBlockSerializerV1 = TLBlockSerializerV1;
