// @super-protocol/sdk-js/build/providers/storage/ChunksDownloadDecorator.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DownloadDecorator = exports.createDownloadChunkMethodWithRetry = exports.downloadChunkMethod = void 0;
const stream_1 = require("stream");
function* generateChunks(dataSize, chunkSize, startOffset = BigInt(0)) {
    let chunk = 0;
    let offset = startOffset;
    while (dataSize > offset) {
        yield {
            offset,
            length: chunkSize > dataSize - offset ? Number(dataSize - offset) : chunkSize,
            dataSize,
            chunk: chunk++,
            chunksCount: Math.ceil(Number(dataSize / BigInt(chunkSize))),
        };
        offset += BigInt(chunkSize);
    }
}
class ChunkedReadableTransform extends stream_1.Transform {
    constructor(downloadChunk, params) {
        super({
            objectMode: true,
            highWaterMark: params.concurrency,
        });
        this.downloadChunk = downloadChunk;
    }
    _transform(chunk, encoding, callback) {
        const resultPromise = this.downloadChunk(chunk);
        callback(null, {
            resultPromise,
            chunk,
        });
    }
}
class ChainedChunksReadableTransform extends stream_1.Transform {
    constructor(progressListener) {
        super({
            objectMode: true,
            highWaterMark: 1,
        });
        this.progressListener = progressListener;
    }
    _transform(obj, encoding, callback) {
        obj.resultPromise.then((data) => {
            callback(null, data);
            if (this.progressListener) {
                this.progressListener(Number(obj.chunk.dataSize), Number(obj.chunk.offset) + obj.chunk.length);
            }
        }, (err) => {
            callback(err);
        });
    }
}
const downloadChunkMethod = (provider, objectPath, chunk) => __awaiter(void 0, void 0, void 0, function* () {
    var _a, e_1, _b, _c;
    let position = 0;
    const buffer = Buffer.alloc(chunk.length);
    const dataStream = yield provider.downloadFile(objectPath, {
        offset: Number(chunk.offset),
        length: chunk.length,
    });
    try {
        for (var _d = true, dataStream_1 = __asyncValues(dataStream), dataStream_1_1; dataStream_1_1 = yield dataStream_1.next(), _a = dataStream_1_1.done, !_a;) {
            _c = dataStream_1_1.value;
            _d = false;
            try {
                const data = _c;
                data.copy(buffer, position);
                position += data.length;
                if (position > chunk.length) {
                    throw new Error('Chunk buffer is overflow, read data is more than requested for chunk');
                }
            }
            finally {
                _d = true;
            }
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (!_d && !_a && (_b = dataStream_1.return)) yield _b.call(dataStream_1);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return buffer;
});
exports.downloadChunkMethod = downloadChunkMethod;
const createDownloadChunkMethodWithRetry = (downloadChunkMethod, { retryMaxCount = 3, retryWaitTimeFactory = () => () => 1000, onRetry, } = {}) => {
    return (provider, objectPath, chunk) => __awaiter(void 0, void 0, void 0, function* () {
        const retryWaitTime = retryWaitTimeFactory();
        //let lastError: unknown = null;
        let retryCount = retryMaxCount;
        while (retryCount > 0) {
            try {
                return yield downloadChunkMethod(provider, objectPath, chunk);
            }
            catch (error) {
                //lastError = error;
            }
            retryCount--;
            if (onRetry) {
                onRetry(null, chunk, retryCount);
            }
            if (retryCount !== 0) {
                yield new Promise((resolve, reject) => setTimeout(resolve, retryWaitTime()));
            }
        }
        throw new Error(`Max retry attempts was reached for object path ${objectPath}, offset ${chunk.offset}, length ${chunk.length
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        }`);
    });
};
exports.createDownloadChunkMethodWithRetry = createDownloadChunkMethodWithRetry;
class DownloadDecorator {
    constructor(provider, chunkDownloaderMethod, options) {
        this.provider = provider;
        this.chunkDownloaderMethod = chunkDownloaderMethod;
        this.options = options;
    }
    uploadFile(inputStream, remotePath, contentLength, progressListener) {
        return this.provider.uploadFile(inputStream, remotePath, contentLength, progressListener);
    }
    downloadFile(remotePath, config, progressListener) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            if (config === null || config === void 0 ? void 0 : config.length) {
                return this.provider.downloadFile(remotePath, config, progressListener);
            }
            const offset = BigInt((_a = this.options.offset) !== null && _a !== void 0 ? _a : 0);
            const objectSize = yield this.provider.getObjectSize(remotePath);
            const chunksStream = stream_1.Readable.from(generateChunks(BigInt(objectSize), this.options.chunkSize, offset));
            const downloadStream = new ChunkedReadableTransform((chunk) => __awaiter(this, void 0, void 0, function* () {
                return this.chunkDownloaderMethod(this.provider, remotePath, chunk);
            }), {
                concurrency: this.options.concurrency,
            });
            const chainedStream = new ChainedChunksReadableTransform(progressListener);
            return chunksStream
                .pipe(downloadStream)
                .on('error', (error) => {
                chainedStream.emit('error', error);
            })
                .pipe(chainedStream);
        });
    }
    deleteObject(remotePath) {
        return this.provider.deleteObject(remotePath);
    }
    listObjects(remotePath) {
        return this.provider.listObjects(remotePath);
    }
    getObjectSize(remotePath) {
        return this.provider.getObjectSize(remotePath);
    }
    getLastModified(remotePath) {
        return this.provider.getLastModified(remotePath);
    }
}
exports.DownloadDecorator = DownloadDecorator;
