// @super-protocol/sdk-js/build/providers/storage/S3StorageProvider.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3StorageProvider = void 0;
const aws_sdk_1 = require("aws-sdk");
class S3StorageProvider {
    constructor(storageAccess) {
        this.multipartChunkSizeInBytes = 64 * 1024 * 1024; // 64MB
        const { accessKeyId, secretAccessKey, endpoint, bucket } = storageAccess;
        if (!accessKeyId)
            throw new Error('Access key id is undefined');
        if (!secretAccessKey)
            throw new Error('Secret access is undefined');
        if (!endpoint)
            throw new Error('Endpoint is undefined');
        if (!bucket)
            throw new Error('Bucket is undefined');
        this.bucket = bucket;
        this.s3Client = new aws_sdk_1.S3({
            credentials: {
                accessKeyId,
                secretAccessKey,
            },
            endpoint,
            s3ForcePathStyle: true,
            signatureVersion: 'v4',
            httpOptions: { timeout: 0 },
        });
    }
    uploadFile(inputStream, remotePath, contentLength, progressListener) {
        return __awaiter(this, void 0, void 0, function* () {
            // For performance & cost optimization
            // https://docs.storj.io/dcs/api-reference/s3-compatible-gateway/multipart-upload/multipart-part-size
            if (inputStream.readableHighWaterMark >= this.multipartChunkSizeInBytes) {
                return this.multipartUpload(inputStream, remotePath, contentLength, progressListener);
            }
            const result = yield this.s3Client
                .putObject({
                Body: inputStream,
                Bucket: this.bucket,
                Key: remotePath,
                ContentLength: contentLength,
            })
                .on('httpUploadProgress', ({ total, loaded }) => progressListener === null || progressListener === void 0 ? void 0 : progressListener(total, loaded))
                .promise();
            if (result.$response.error) {
                throw result.$response.error;
            }
        });
    }
    multipartUpload(inputStream, remotePath, contentLength, progressListener) {
        var _a, inputStream_1, inputStream_1_1;
        var _b, e_1, _c, _d;
        return __awaiter(this, void 0, void 0, function* () {
            const multipart = yield this.s3Client
                .createMultipartUpload({
                Bucket: this.bucket,
                Key: remotePath,
            })
                .promise();
            if (!multipart.UploadId) {
                throw new Error('UploadId property is empty');
            }
            try {
                let totalWritten = 0;
                let partNumber = 0;
                const uploadId = multipart.UploadId;
                const parts = [];
                try {
                    for (_a = true, inputStream_1 = __asyncValues(inputStream); inputStream_1_1 = yield inputStream_1.next(), _b = inputStream_1_1.done, !_b;) {
                        _d = inputStream_1_1.value;
                        _a = false;
                        try {
                            const buffer = _d;
                            partNumber++;
                            const response = yield this.s3Client
                                .uploadPart({
                                Body: buffer,
                                Bucket: this.bucket,
                                Key: remotePath,
                                UploadId: uploadId,
                                PartNumber: partNumber,
                            })
                                .promise();
                            parts.push({
                                ETag: response.ETag,
                                PartNumber: partNumber,
                            });
                            totalWritten += buffer.length;
                            if (!!progressListener) {
                                progressListener(contentLength, totalWritten);
                            }
                        }
                        finally {
                            _a = true;
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (!_a && !_b && (_c = inputStream_1.return)) yield _c.call(inputStream_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
                const doneParams = {
                    Bucket: this.bucket,
                    Key: remotePath,
                    UploadId: multipart.UploadId,
                    MultipartUpload: { Parts: parts },
                };
                yield this.s3Client.completeMultipartUpload(doneParams).promise();
            }
            catch (uploadingError) {
                yield this.s3Client
                    .abortMultipartUpload({
                    Bucket: this.bucket,
                    Key: remotePath,
                    UploadId: multipart.UploadId,
                })
                    .promise();
                throw uploadingError;
            }
        });
    }
    deleteObject(remotePath) {
        return __awaiter(this, void 0, void 0, function* () {
            const deleteResult = yield this.s3Client
                .deleteObject({
                Bucket: this.bucket,
                Key: remotePath,
            })
                .promise();
            if (deleteResult.$response.error) {
                throw deleteResult.$response.error;
            }
        });
    }
    downloadFile(remotePath, config, progressListener) {
        return __awaiter(this, void 0, void 0, function* () {
            const getObjectParams = {
                Bucket: this.bucket,
                Key: remotePath,
            };
            if (config) {
                const start = config.offset || 0;
                const end = start + (config.length || 0);
                getObjectParams.Range = `bytes=${start}-${end || ''}`;
            }
            const downloadStream = this.s3Client.getObject(getObjectParams).createReadStream();
            let current = 0;
            if (progressListener) {
                const fileBytesSize = config.length || (yield this.getObjectSize(remotePath));
                downloadStream.on('data', (chunk) => {
                    current += chunk.length;
                    progressListener(fileBytesSize, current);
                });
            }
            return downloadStream;
        });
    }
    listObjects(remotePath) {
        return __awaiter(this, void 0, void 0, function* () {
            const prefix = remotePath.endsWith('/') ? remotePath : `${remotePath}/`;
            const listObjects = yield this.s3Client
                .listObjectsV2({
                Bucket: this.bucket,
                Prefix: prefix,
            })
                .promise();
            if (listObjects.$response.error) {
                throw listObjects.$response.error;
            }
            let result = [];
            if (listObjects.Contents) {
                result = listObjects.Contents.map((object) => ({
                    name: object.Key || '',
                    createdAt: object.LastModified || new Date(),
                    size: object.Size || 0,
                }));
            }
            return result;
        });
    }
    getMetadata(remotePath) {
        return __awaiter(this, void 0, void 0, function* () {
            const getObjectParams = {
                Bucket: this.bucket,
                Key: remotePath,
            };
            const metadata = yield this.s3Client.headObject(getObjectParams).promise();
            if (metadata.$response.error) {
                throw metadata.$response.error;
            }
            return metadata;
        });
    }
    getObjectSize(remotePath) {
        return __awaiter(this, void 0, void 0, function* () {
            const metadata = yield this.getMetadata(remotePath);
            if (metadata.ContentLength === undefined) {
                throw new Error('ContentLength property is empty');
            }
            return metadata.ContentLength || 0;
        });
    }
    getLastModified(remotePath) {
        return __awaiter(this, void 0, void 0, function* () {
            const metadata = yield this.getMetadata(remotePath);
            if (!metadata.LastModified) {
                throw new Error('LastModified property is empty');
            }
            return metadata.LastModified;
        });
    }
}
exports.S3StorageProvider = S3StorageProvider;
