// @super-protocol/sdk-js/build/providers/storage/StorageKeyValueAdapter.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const stream_1 = require("stream");
const dto_js_1 = require("@super-protocol/dto-js");
const getStorageProvider_1 = __importDefault(require("./getStorageProvider"));
const logger_1 = __importDefault(require("../../logger"));
const Crypto_1 = __importDefault(require("../../crypto/Crypto"));
class StorageKeyValueAdapter {
    constructor(storageAccess, config) {
        if (!(storageAccess === null || storageAccess === void 0 ? void 0 : storageAccess.credentials))
            throw new Error('Credentials is empty');
        const { showLogs = true } = config || {};
        this.logger = showLogs ? logger_1.default.child({ class: StorageKeyValueAdapter.name }) : null;
        this.storageProvider = (0, getStorageProvider_1.default)(storageAccess);
    }
    decrypt(encryption, key) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!encryption)
                return null;
            if (!key)
                throw new Error('Key cannot be empty!');
            encryption.key = key;
            return JSON.parse(yield Crypto_1.default.decrypt(encryption));
        });
    }
    encrypt(data, key) {
        return __awaiter(this, void 0, void 0, function* () {
            if (data === undefined)
                throw new Error('Data cannot be empty!');
            if (!key)
                throw new Error('Private cannot be empty!');
            return Crypto_1.default.encrypt(JSON.stringify(data), {
                algo: dto_js_1.CryptoAlgorithm.AES,
                encoding: dto_js_1.Encoding.base64,
                key,
                cipher: dto_js_1.Cipher.AES_256_GCM,
            });
        });
    }
    downloadFromStorage(filepath) {
        return __awaiter(this, void 0, void 0, function* () {
            const downloadStream = yield this.storageProvider.downloadFile(filepath, {});
            return this.streamToString(downloadStream);
        });
    }
    streamToString(stream) {
        const chunks = [];
        return new Promise((resolve, reject) => {
            stream.on('data', (chunk) => {
                chunks.push(Buffer.from(chunk));
            });
            stream.on('error', (err) => reject(err));
            stream.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')));
        });
    }
    storageUpload(key, value, privateKey) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const encryptedValue = yield this.encrypt(value, privateKey);
                const buffer = Buffer.from(JSON.stringify(encryptedValue));
                yield this.storageProvider.uploadFile(stream_1.Readable.from(buffer), key, buffer.byteLength);
                (_a = this.logger) === null || _a === void 0 ? void 0 : _a.info({ data: key }, 'Success uploading to storage');
            }
            catch (err) {
                (_b = this.logger) === null || _b === void 0 ? void 0 : _b.error({ err }, 'Error uploading to storage');
                throw err;
            }
        });
    }
    storageDelete(key) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.storageProvider.deleteObject(key);
                (_a = this.logger) === null || _a === void 0 ? void 0 : _a.info({ data: key }, 'Success deleting from storage');
            }
            catch (err) {
                (_b = this.logger) === null || _b === void 0 ? void 0 : _b.info({ err }, 'Error deleting from storage');
                throw err;
            }
        });
    }
    storageDownload(key, privateKey) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const downloaded = yield this.downloadFromStorage(key);
                (_a = this.logger) === null || _a === void 0 ? void 0 : _a.info({ key }, 'Success download data from storage');
                if (!downloaded)
                    return null;
                const decryptedValue = yield this.decrypt(JSON.parse(downloaded), privateKey);
                return decryptedValue;
            }
            catch (err) {
                (_b = this.logger) === null || _b === void 0 ? void 0 : _b.info({
                    err,
                    key,
                }, 'Error download data from storage');
                throw err;
            }
        });
    }
    storageListFiles(key) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const listObjects = yield this.storageProvider.listObjects(key);
                (_a = this.logger) === null || _a === void 0 ? void 0 : _a.trace({
                    data: listObjects,
                    key,
                }, 'Success list objects from storage');
                return listObjects.filter((obj) => !obj.isFolder);
            }
            catch (err) {
                (_b = this.logger) === null || _b === void 0 ? void 0 : _b.info({
                    err,
                    key,
                }, 'Error list objects from storage');
                throw err;
            }
        });
    }
    set(key, value, privateKey) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.storageUpload(key, value, privateKey);
        });
    }
    delete(key) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.storageDelete(key);
        });
    }
    get(key, privateKey) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.storageDownload(key, privateKey);
        });
    }
    listFiles(key) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.storageListFiles(key);
        });
    }
}
exports.default = StorageKeyValueAdapter;
