// @super-protocol/sdk-js/build/providers/storage/StorjStorageProvider.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
    function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const buffer_1 = require("buffer");
const helper_1 = require("../../utils/helper");
const stream_1 = __importDefault(require("stream"));
const logger_1 = __importDefault(require("../../logger"));
class StorJStorageProvider {
    constructor(credentials, maximumConcurrent) {
        this.logger = logger_1.default.child({ className: 'StorJStorageProvider' });
        if (!(0, helper_1.isNodeJS)()) {
            throw Error('StorageProvider: StorJ is supported only in the node.js execution environment');
        }
        this.bucket = '';
        this.prefix = '';
        if (credentials.bucket) {
            this.bucket = credentials.bucket;
            this.prefix = credentials.prefix;
        }
        else if (credentials.storageId) {
            // back compatibility
            this.bucket = credentials.storageId;
        }
        this.accessToken = credentials.token;
        this.maximumConcurrent = maximumConcurrent;
    }
    uploadFile(inputStream, remotePath, contentLength, progressListener) {
        var _a, inputStream_1, inputStream_1_1;
        var _b, e_1, _c, _d;
        return __awaiter(this, void 0, void 0, function* () {
            const storj = yield this.lazyStorj();
            const options = new storj.UploadOptions();
            const project = yield this.lazyProject();
            const uploader = yield project.uploadObject(this.bucket, this.prefix + remotePath, options);
            let totalWritten = 0;
            try {
                try {
                    for (_a = true, inputStream_1 = __asyncValues(inputStream); inputStream_1_1 = yield inputStream_1.next(), _b = inputStream_1_1.done, !_b;) {
                        _d = inputStream_1_1.value;
                        _a = false;
                        try {
                            const buffer = _d;
                            yield uploader.write(buffer, buffer.length);
                            totalWritten += buffer.length;
                            if (!!progressListener) {
                                progressListener(contentLength, totalWritten);
                            }
                        }
                        finally {
                            _a = true;
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (!_a && !_b && (_c = inputStream_1.return)) yield _c.call(inputStream_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
                yield uploader.commit();
            }
            catch (uploadingError) {
                try {
                    yield uploader.abort();
                }
                catch (abortingError) {
                    logger_1.default.error({ err: abortingError }, 'Failed to abort file uploading');
                }
                throw uploadingError;
            }
        });
    }
    downloadFile(remotePath, config, progressListener) {
        return __awaiter(this, void 0, void 0, function* () {
            const storj = yield this.lazyStorj();
            const project = yield this.lazyProject();
            const length = config.length || (yield this.getObjectSize(remotePath));
            const options = new storj.DownloadOptions(config.offset || 0, length);
            const downloader = yield project.downloadObject(this.bucket, this.prefix + remotePath, options);
            const loader = function () {
                return __asyncGenerator(this, arguments, function* () {
                    const readBuffer = buffer_1.Buffer.alloc(StorJStorageProvider.DOWNLOAD_BUFFER_SIZE);
                    let current = 0;
                    while (current < length) {
                        // We have to cast result to any, because of the wrong type declartion in uplink-nodejs.
                        const downloadResult = yield __await(downloader.read(readBuffer, readBuffer.length));
                        const bytesRead = downloadResult.bytes_read;
                        current += bytesRead;
                        yield yield __await(buffer_1.Buffer.from(readBuffer.subarray(0, bytesRead)));
                        if (!!progressListener) {
                            progressListener(length, current);
                        }
                    }
                });
            };
            return stream_1.default.Readable.from(loader()).on('close', () => __awaiter(this, void 0, void 0, function* () {
                yield downloader.close();
            }));
        });
    }
    deleteObject(remotePath) {
        return __awaiter(this, void 0, void 0, function* () {
            const project = yield this.lazyProject();
            yield project.deleteObject(this.bucket, this.prefix + remotePath);
        });
    }
    listObjects(remotePath) {
        return __awaiter(this, void 0, void 0, function* () {
            const storj = yield this.lazyStorj();
            const project = yield this.lazyProject();
            const objects = yield project.listObjects(this.bucket, {
                recursive: true,
                cursor: '',
                prefix: this.prefix + remotePath,
                system: true,
                custom: true,
            });
            const result = [];
            for (const key in Object.keys(objects)) {
                const value = objects[key];
                result.push({
                    name: value.key,
                    size: value.system.content_length,
                    isFolder: value.is_prefix == 1,
                    childrenCount: value.custom.count,
                    createdAt: new Date(value.system.created * 1000), // TODO: check timezone
                });
            }
            return result;
        });
    }
    getObjectSize(remotePath) {
        return __awaiter(this, void 0, void 0, function* () {
            const project = yield this.lazyProject();
            const objectInfo = yield project.statObject(this.bucket, this.prefix + remotePath);
            return objectInfo.system.content_length;
        });
    }
    getLastModified(remotePath) {
        return __awaiter(this, void 0, void 0, function* () {
            const project = yield this.lazyProject();
            const objectInfo = yield project.statObject(this.bucket, this.prefix + remotePath);
            return new Date(objectInfo.system.created * 1000);
        });
    }
    lazyStorj() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this._storj) {
                this._storj = yield require('@super-protocol/uplink-nodejs');
            }
            return this._storj;
        });
    }
    lazyAccess() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this._access) {
                const storj = yield this.lazyStorj();
                const uplink = new storj.Uplink();
                this._access = yield uplink.parseAccess(this.accessToken);
            }
            return this._access;
        });
    }
    lazyProject() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this._project) {
                const access = yield this.lazyAccess();
                this._project = yield access.openProject(this.maximumConcurrent);
            }
            return this._project;
        });
    }
}
exports.default = StorJStorageProvider;
StorJStorageProvider.DOWNLOAD_BUFFER_SIZE = 4194304; // 4mb
