// @super-protocol/sdk-js/build/providers/storage/StorjAdapter.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const logger_1 = __importDefault(require("../../logger"));
const StorageAdapter_1 = __importDefault(require("./StorageAdapter"));
class StorjAdapter {
    constructor(storageAccess, config) {
        const { showLogs = true } = config || {};
        this.logger = showLogs ? logger_1.default.child({ class: StorjAdapter.name }) : null;
        this.storageAdapter = new StorageAdapter_1.default(storageAccess, config);
        this.storageAdapter.run();
    }
    subscribe(cb) {
        var _a;
        return (_a = this.storageAdapter) === null || _a === void 0 ? void 0 : _a.subscribe(cb);
    }
    get(key, encryptionKey) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.storageAdapter.get(key, encryptionKey).catch((err) => {
                var _a, _b;
                const message = ((_a = err.message) === null || _a === void 0 ? void 0 : _a.toLowerCase()) || '';
                if (message.includes('object not found') || message.includes('object has been deleted')) {
                    (_b = this.logger) === null || _b === void 0 ? void 0 : _b.info({ key }, 'Object not found');
                    return null;
                }
                throw err;
            });
        });
    }
    has(key) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.storageAdapter.has(key);
        });
    }
    set(key, value, encryptionKey) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.storageAdapter.set(key, value, encryptionKey).catch((err) => {
                var _a, _b;
                const message = ((_a = err.message) === null || _a === void 0 ? void 0 : _a.toLowerCase()) || '';
                if (message.includes('object has been deleted')) {
                    (_b = this.logger) === null || _b === void 0 ? void 0 : _b.info({ key }, 'Object has been deleted');
                    return;
                }
                throw err;
            });
        });
    }
    del(key) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.storageAdapter.delete(key);
        });
    }
    stop() {
        this.storageAdapter.stop();
    }
    shutdown() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.storageAdapter.shutdown();
        });
    }
}
exports.default = StorjAdapter;
