// @super-protocol/sdk-js/build/providers/storage/StorageMetadataReader.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const logger_1 = __importDefault(require("../../logger"));
class StorageMetadataReader {
    constructor(config) {
        const { showLogs = true, objectDeletedFlag, storageKeyValueAdapter } = config;
        this.logger = showLogs ? logger_1.default.child({ class: StorageMetadataReader.name }) : null;
        this.storageKeyValueAdapter = storageKeyValueAdapter;
        this.objectDeletedFlag = objectDeletedFlag;
    }
    listInstances(key) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.storageKeyValueAdapter
                .listFiles(`${key}/`)
                .then((objects) => new Map(objects.map((obj) => { var _a; return [((_a = obj.name.split('/')) === null || _a === void 0 ? void 0 : _a.pop()) || obj.name, obj]; })));
        });
    }
    fetchInstancesUpdates(key, currentInstances) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            const result = {
                updated: new Map(),
                deleted: new Set(),
            };
            try {
                const directoryInstances = yield this.listInstances(key);
                if (directoryInstances.has(this.objectDeletedFlag)) {
                    result.deleted.add(key);
                    return result;
                }
                directoryInstances.forEach((instance, instanceId) => {
                    const currentInstance = currentInstances.get(instanceId);
                    if (!currentInstance || currentInstance.modifiedTs < instance.createdAt.getTime()) {
                        result.updated.set(instanceId, instance);
                    }
                });
                currentInstances.forEach((_, instanceId) => {
                    if (!directoryInstances.has(instanceId)) {
                        result.deleted.add(instanceId);
                    }
                });
                (_a = this.logger) === null || _a === void 0 ? void 0 : _a.trace({
                    updated: result.updated.size,
                    deleted: result.deleted.size,
                }, 'Check result');
                return result;
            }
            catch (error) {
                (_b = this.logger) === null || _b === void 0 ? void 0 : _b.error({ error }, 'Error fetching remote copy');
                return result;
            }
        });
    }
}
exports.default = StorageMetadataReader;
