// @super-protocol/sdk-js/build/crypto/Crypto.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const dto_js_1 = require("@super-protocol/dto-js");
const AES_1 = __importDefault(require("./nodejs/AES"));
const ARIA_1 = __importDefault(require("./nodejs/ARIA"));
const ECIES_1 = __importDefault(require("./nodejs/ECIES"));
const RSA_Hybrid_1 = __importDefault(require("./nodejs/RSA-Hybrid"));
const NativeCrypto_1 = __importDefault(require("./nodejs/NativeCrypto"));
class Crypto {
    /**
     * Used to encrypt data before sending it to blockchain
     * @param algorithm - encryption algorithm
     * @param content - string data to encrypt
     * @param key - key in string format (default encoding base64)
     * @param encoding - encoding of key and output data
     * @returns {Promise<Encryption>} - object what contains encrypted data, key and spec to decryption
     */
    static encrypt(content, encryption) {
        return __awaiter(this, void 0, void 0, function* () {
            switch (encryption.algo) {
                case dto_js_1.CryptoAlgorithm.AES:
                    return yield AES_1.default.encrypt(content, encryption);
                case dto_js_1.CryptoAlgorithm.ARIA:
                    return yield ARIA_1.default.encrypt(content, encryption);
                case dto_js_1.CryptoAlgorithm.ECIES:
                    return yield ECIES_1.default.encrypt(content, encryption);
                case dto_js_1.CryptoAlgorithm.RSAHybrid:
                    return yield RSA_Hybrid_1.default.encrypt(content, encryption);
                default:
                    throw Error(`${encryption.algo} algorithm not supported`);
            }
        });
    }
    /**
     * Encrypts data stream
     * @param inputStream - path to file that will be encrypted
     * @param outputStream - place where it will be saved
     * @param algorithm - file encryption algorithm
     * @param key – key that will be used to encrypt data
     * @returns {Promise<Encryption>} - encryption info
     */
    static encryptStream(inputStream, outputStream, encryption) {
        return __awaiter(this, void 0, void 0, function* () {
            switch (encryption.algo) {
                case dto_js_1.CryptoAlgorithm.AES:
                    return yield AES_1.default.encryptStream(inputStream, outputStream, encryption);
                case dto_js_1.CryptoAlgorithm.ARIA:
                    return yield ARIA_1.default.encryptStream(inputStream, outputStream, encryption);
                case dto_js_1.CryptoAlgorithm.RSAHybrid:
                    return yield RSA_Hybrid_1.default.encryptStream(inputStream, outputStream, encryption);
                default:
                    throw Error(`${encryption.algo} algorithm not supported`);
            }
        });
    }
    /**
     * Used to decrypt data from blockchain
     * @param encryption - object what contains encrypted data, key and spec to decryption
     * @returns {Promise<string>} - decrypted string
     */
    static decrypt(encryption) {
        return __awaiter(this, void 0, void 0, function* () {
            switch (encryption.algo) {
                case dto_js_1.CryptoAlgorithm.AES:
                    return AES_1.default.decrypt(encryption);
                case dto_js_1.CryptoAlgorithm.ARIA:
                    return ARIA_1.default.decrypt(encryption);
                case dto_js_1.CryptoAlgorithm.ECIES:
                    return yield ECIES_1.default.decrypt(encryption);
                case dto_js_1.CryptoAlgorithm.RSAHybrid:
                    return RSA_Hybrid_1.default.decrypt(encryption);
                default:
                    throw Error(`${encryption.algo} algorithm not supported`);
            }
        });
    }
    /**
     * Decrypts data stream
     * @param inputStream - stream with data to decrypt
     * @param outputStream - stream where the decrypted data will be written
     * @param encryption – encryption info
     */
    static decryptStream(inputStream, outputStream, encryption) {
        return __awaiter(this, void 0, void 0, function* () {
            switch (encryption.algo) {
                case dto_js_1.CryptoAlgorithm.AES:
                    return yield AES_1.default.decryptStream(inputStream, outputStream, encryption);
                case dto_js_1.CryptoAlgorithm.ARIA:
                    return yield ARIA_1.default.decryptStream(inputStream, outputStream, encryption);
                case dto_js_1.CryptoAlgorithm.RSAHybrid:
                    return yield RSA_Hybrid_1.default.decryptStream(inputStream, outputStream, encryption);
                default:
                    throw Error(`${encryption.algo} algorithm not supported`);
            }
        });
    }
    static createHash(param1, hashInfo) {
        return __awaiter(this, void 0, void 0, function* () {
            const { algo, encoding } = hashInfo;
            return Buffer.isBuffer(param1)
                ? NativeCrypto_1.default.createHashFromBuffer(param1, algo, encoding)
                : yield NativeCrypto_1.default.createHashFromStream(param1, algo, encoding);
        });
    }
}
exports.default = Crypto;
