// @super-protocol/tee-lib/dist/tee-device/zero-tee-device.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZeroTeeDeviceService = void 0;
const calculate_hash_1 = require("../helpers/calculate-hash");
/**
 * For testing purpose only.
 */
class ZeroTeeDeviceService {
    constructor(params) {
        this.logger = params.logger?.child({ name: ZeroTeeDeviceService.name });
    }
    parseAndValidateTlsCertificate() {
        this.logger?.trace({ method: 'parseAndValidateTlsCertificate' }, 'Get');
        const result = {
            dataHash: Buffer.from('000102030405060708090a0b0c0d0f', 'hex'),
            mrEnclave: Buffer.from('000102030405060708090a0b0c0d0f000102030405060708090a0b0c0d0f0102', 'hex'),
            mrSigner: Buffer.from('000102030405060708090a0b0c0d0f000102030405060708090a0b0c0d0f0102', 'hex'),
            userData: Buffer.from('000102030405060708090a0b0c0d0f', 'hex'),
        };
        return Promise.resolve(result);
    }
    getUniqKey(salt) {
        this.logger?.trace({ method: 'getSharedUniqKey' }, 'Get');
        return Promise.resolve((0, calculate_hash_1.calculateHash)(Buffer.concat([Buffer.alloc(16), salt])));
    }
    getDeviceId() {
        this.logger?.trace({ method: 'getDeviceId' }, 'Get');
        return Promise.resolve(Buffer.from('0102030405060708090a0b0c0d0e0f10', 'hex'));
    }
    async getDataOf(userData) {
        const logger = this.logger?.child({ method: 'getDataOf' });
        logger?.trace({ userData: userData.slice(0, 64).toString('hex') }, 'Return');
        return Buffer.concat([
            Buffer.concat([userData, Buffer.alloc(64)]).slice(0, 64),
            Buffer.concat([await this.getDeviceId(), Buffer.alloc(16)]).slice(0, 16),
            Buffer.alloc(1024),
            // randomBytes(16),
        ]);
    }
}
exports.ZeroTeeDeviceService = ZeroTeeDeviceService;
