// @super-protocol/sdk-js/build/types/Consensus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConsensusConstantsStructure = exports.TcbStatus = exports.TcbVerifiedStatus = void 0;
var TcbVerifiedStatus;
(function (TcbVerifiedStatus) {
    TcbVerifiedStatus[TcbVerifiedStatus["Valid"] = 0] = "Valid";
    TcbVerifiedStatus[TcbVerifiedStatus["InvalidQuote"] = 1] = "InvalidQuote";
    TcbVerifiedStatus[TcbVerifiedStatus["InvalidMrEnclave"] = 2] = "InvalidMrEnclave";
    TcbVerifiedStatus[TcbVerifiedStatus["InvalidBcbHash"] = 3] = "InvalidBcbHash";
})(TcbVerifiedStatus = exports.TcbVerifiedStatus || (exports.TcbVerifiedStatus = {}));
var TcbStatus;
(function (TcbStatus) {
    TcbStatus["Inited"] = "0";
    TcbStatus["Completed"] = "1";
    TcbStatus["Banned"] = "2";
    TcbStatus["BenchmarkChanged"] = "3";
})(TcbStatus = exports.TcbStatus || (exports.TcbStatus = {}));
exports.ConsensusConstantsStructure = {
    CONSENSUS_L1: Number,
    CONSENSUS_L2: Number,
    CONSENSUS_K: Number,
    CONSENSUS_MAX_PENALTIES: Number,
    MAX_BECHMARK_OVERRUN_PERCENT: Number,
    CONSENSUS_EPOCH_DURATION: Number,
};
