// @super-protocol/sdk-js/build/staticModels/ModelPackager.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModelPackager = void 0;
const Compression_1 = require("../proto/Compression");
const compressors_1 = require("../utils/compressors");
class ModelPackager {
    /**
     * Compressing message with best algorithm
     *
     * @param message - message
     * @returns Compressed message in bytes representation
     */
    static pack(message) {
        return __awaiter(this, void 0, void 0, function* () {
            const promises = ModelPackager.compressors.map((Compressor) => __awaiter(this, void 0, void 0, function* () {
                const compressorInstance = new Compressor();
                const data = yield compressorInstance.compress(message);
                return {
                    compressionType: Compressor.type,
                    data,
                    size: data.byteLength,
                };
            }));
            const options = yield Promise.all(promises);
            options.sort((first, second) => first.size - second.size);
            const minimal = options[0];
            const encoded = Compression_1.Compression.encode({
                type: minimal.compressionType,
                data: minimal.data,
            }).finish();
            return Buffer.from(encoded);
        });
    }
    /**
     * Uncompressing message
     *
     * @param message - message with bytes representation
     * @returns original message
     */
    static unpack(message) {
        return __awaiter(this, void 0, void 0, function* () {
            const decoded = Compression_1.Compression.decode(message);
            const Compressor = ModelPackager.compressors.find((CompressorClass) => CompressorClass.type === decoded.type);
            if (!Compressor) {
                throw new Error(`Cannot unpack message. Compressor ${decoded.type} isn't supported`);
            }
            return new Compressor().decompress(Buffer.from(decoded.data));
        });
    }
}
exports.ModelPackager = ModelPackager;
ModelPackager.compressors = [compressors_1.GzipCompressor, compressors_1.UncompressedCompressor];
