// @super-protocol/sdk-js/build/staticModels/ProviderRegistry.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const logger_1 = __importDefault(require("../logger"));
const helper_1 = require("../utils/helper");
const connectors_1 = require("../connectors");
const TxManager_1 = __importDefault(require("../utils/TxManager"));
class ProviderRegistry {
    /**
     * Function for fetching list of all providers addresses
     */
    static getAllProviders() {
        return __awaiter(this, void 0, void 0, function* () {
            const contract = connectors_1.BlockchainConnector.getInstance().getContract();
            this.providers = yield contract.methods.getProvidersAuths().call();
            return this.providers;
        });
    }
    /**
     * Fetch provider security deposit by provider authority account
     */
    static getSecurityDeposit(providerAuthority) {
        return __awaiter(this, void 0, void 0, function* () {
            const contract = connectors_1.BlockchainConnector.getInstance().getContract();
            return (0, helper_1.convertBigIntToString)(yield contract.methods.getProviderSecurityDeposit(providerAuthority).call());
        });
    }
    static isProviderRegistered(providerAuthority) {
        const contract = connectors_1.BlockchainConnector.getInstance().getContract();
        return contract.methods.isProviderRegistered(providerAuthority).call();
    }
    /**
     * Refills security provider deposit
     * Call this function with provider authority account (in transactionOptions)
     * @param amount - amount of additional tokens
     * @param recipient - target provider authority address
     * @param transactionOptions - object what contains alternative action account or gas limit (optional)
     */
    static refillSecurityDepositFor(amount, recipient, transactionOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const contract = connectors_1.BlockchainConnector.getInstance().getContract();
            (0, helper_1.checkIfActionAccountInitialized)(transactionOptions);
            yield TxManager_1.default.execute(contract.methods.refillProviderSecurityDepoFor(recipient, amount), transactionOptions);
        });
    }
    /**
     * Reg new provider
     * @param providerInfo - data of new provider
     * @param transactionOptions - object what contains alternative action account or gas limit (optional)
     */
    static registerProvider(providerInfo, transactionOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const contract = connectors_1.BlockchainConnector.getInstance().getContract();
            (0, helper_1.checkIfActionAccountInitialized)(transactionOptions);
            yield TxManager_1.default.execute(contract.methods.registerProvider(providerInfo), transactionOptions);
        });
    }
    /**
     * Refills security deposit for provider
     * Call this function with provider authority account (in transactionOptions)
     * @param amount - amount of additional tokens
     * @param transactionOptions - object what contains alternative action account or gas limit (optional)
     */
    static refillSecurityDeposit(amount, transactionOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const contract = connectors_1.BlockchainConnector.getInstance().getContract();
            (0, helper_1.checkIfActionAccountInitialized)(transactionOptions);
            yield TxManager_1.default.execute(contract.methods.refillProviderSecurityDepo(amount), transactionOptions);
        });
    }
    /**
     * Return security deposit for provider
     * Call this function with provider authority account (in transactionOptions)
     * @param amount - amount of tokens to return
     * @param transactionOptions - object what contains alternative action account or gas limit (optional)
     */
    static returnSecurityDeposit(amount, transactionOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const contract = connectors_1.BlockchainConnector.getInstance().getContract();
            (0, helper_1.checkIfActionAccountInitialized)(transactionOptions);
            yield TxManager_1.default.execute(contract.methods.returnProviderSecurityDepo(amount), transactionOptions);
        });
    }
    /**
     * Function for adding event listeners on provider registered event in provider registry
     * @param callback - function for processing new provider
     * @returns unsubscribe - unsubscribe function from event
     */
    static onProviderRegistered(callback) {
        const contract = connectors_1.BlockchainEventsListener.getInstance().getContract();
        const logger = this.logger.child({ method: 'onProviderRegistered' });
        const subscription = contract.events.ProviderRegistered();
        subscription.on('data', (event) => {
            callback(event.returnValues.auth, {
                index: Number(event.blockNumber),
                hash: event.blockHash,
            });
        });
        subscription.on('error', (error) => {
            logger.warn(error);
        });
        return () => subscription.unsubscribe();
    }
    /**
     * Function for adding event listeners on provider modified event in provider registry
     * @param callback - function for processing modified provider
     * @returns unsubscribe - unsubscribe function from event
     */
    static onProviderModified(callback) {
        const contract = connectors_1.BlockchainEventsListener.getInstance().getContract();
        const logger = this.logger.child({ method: 'onProviderModified' });
        const subscription = contract.events.ProviderModified();
        subscription.on('data', (event) => {
            callback(event.returnValues.auth, {
                index: Number(event.blockNumber),
                hash: event.blockHash,
            });
        });
        subscription.on('error', (error) => {
            logger.warn(error);
        });
        return () => subscription.unsubscribe();
    }
    /**
     * Function for adding event listeners on provider violation rate incremented event in provider registry
     * @param callback - function for processing new violation rate
     * @returns unsubscribe - unsubscribe function from event
     */
    static onProviderViolationRateIncremented(callback) {
        const contract = connectors_1.BlockchainEventsListener.getInstance().getContract();
        const logger = this.logger.child({ method: 'onProviderViolationRateIncremented' });
        const subscription = contract.events.ProviderViolationRateIncremented();
        subscription.on('data', (event) => {
            const parsedEvent = (0, helper_1.cleanWeb3Data)(event.returnValues);
            callback(parsedEvent.auth, parsedEvent.newViolationRate, {
                index: Number(event.blockNumber),
                hash: event.blockHash,
            });
        });
        subscription.on('error', (error) => {
            logger.warn(error);
        });
        return () => subscription.unsubscribe();
    }
    /**
     * Function for adding event listeners on provider security deposit refilled event in provider registry
     * @param callback - function for processing refilled security deposit
     * @returns unsubscribe - unsubscribe function from event
     */
    static onProviderSecurityDepoRefilled(callback) {
        const contract = connectors_1.BlockchainEventsListener.getInstance().getContract();
        const logger = this.logger.child({ method: 'onProviderSecurityDepoRefilled' });
        const subscription = contract.events.ProviderSecurityDepoRefilled();
        subscription.on('data', (event) => {
            const parsedEvent = (0, helper_1.cleanWeb3Data)(event.returnValues);
            callback(parsedEvent.auth, parsedEvent.amount, {
                index: Number(event.blockNumber),
                hash: event.blockHash,
            });
        });
        subscription.on('error', (error) => {
            logger.warn(error);
        });
        return () => subscription.unsubscribe();
    }
    /**
     * Function for adding event listeners on provider security deposit unlocked event in provider registry
     * @param callback - function for processing unlocked security deposit
     * @returns unsubscribe - unsubscribe function from event
     */
    static onProviderSecurityDepoUnlocked(callback) {
        const contract = connectors_1.BlockchainEventsListener.getInstance().getContract();
        const logger = this.logger.child({ method: 'onProviderSecurityDepoUnlocked' });
        const subscription = contract.events.ProviderSecurityDepoUnlocked();
        subscription.on('data', (event) => {
            const parsedEvent = (0, helper_1.cleanWeb3Data)(event.returnValues);
            callback(parsedEvent.auth, parsedEvent.amount, {
                index: Number(event.blockNumber),
                hash: parsedEvent.blockHash,
            });
        });
        subscription.on('error', (error) => {
            logger.warn(error);
        });
        return () => subscription.unsubscribe();
    }
}
ProviderRegistry.logger = logger_1.default.child({ className: 'ProviderRegistry' });
exports.default = ProviderRegistry;
