// @super-protocol/sdk-js/build/staticModels/SuperproToken.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const logger_1 = __importDefault(require("../logger"));
const abi_1 = require("../contracts/abi");
const store_1 = __importDefault(require("../store"));
const helper_1 = require("../utils/helper");
const TxManager_1 = __importDefault(require("../utils/TxManager"));
class SuperproToken {
    static get addressHttps() {
        return SuperproToken._addressHttps;
    }
    static set addressHttps(newAddress) {
        SuperproToken._addressHttps = newAddress;
        SuperproToken.contractHttps = new store_1.default.web3Https.eth.Contract(abi_1.abi, newAddress, {
            provider: store_1.default.web3Https.currentProvider,
            config: { contractDataInputFill: 'data' },
        });
    }
    static get addressWss() {
        return SuperproToken._addressWss;
    }
    static set addressWss(newAddress) {
        SuperproToken._addressWss = newAddress;
        SuperproToken.contractWss = new store_1.default.web3Wss.eth.Contract(abi_1.abi, newAddress, {
            provider: store_1.default.web3Wss.currentProvider,
            config: { contractDataInputFill: 'data' },
        });
    }
    /**
     * Checks if contract has been initialized, if not - initialize contract
     */
    static checkInit() {
        if (!SuperproToken.contractHttps) {
            throw Error(`SuperproToken must be initialized before it can be used`);
        }
        return SuperproToken.contractHttps;
    }
    /**
     * Checks if contract has been initialized with socket support
     */
    static checkWssInit() {
        return SuperproToken.contractWss;
    }
    /**
     * Fetching balance of SuperProtocol tokens on address
     */
    static balanceOf(address) {
        return __awaiter(this, void 0, void 0, function* () {
            this.checkInit();
            return (0, helper_1.convertBigIntToString)(yield this.contractHttps.methods.balanceOf(address).call());
        });
    }
    /**
     * Fetching allowance of SuperProtocol tokens on address
     */
    static allowance(from, to) {
        return __awaiter(this, void 0, void 0, function* () {
            this.checkInit();
            return (0, helper_1.convertBigIntToString)(yield this.contractHttps.methods.allowance(from, to).call());
        });
    }
    /**
     * Transfers specific amount of SP tokens to specific address
     * @param to - address to revive tokens
     * @param amount - amount of tokens to transfer
     * @param transactionOptions - object what contains alternative action account or gas limit (optional)
     */
    static transfer(to, amount, transactionOptions, checkTxBeforeSend = false) {
        return __awaiter(this, void 0, void 0, function* () {
            const contract = this.checkInit();
            (0, helper_1.checkIfActionAccountInitialized)(transactionOptions);
            if (checkTxBeforeSend) {
                yield TxManager_1.default.dryRun(contract.methods.transfer(to, amount), transactionOptions);
            }
            const receipt = yield TxManager_1.default.execute(contract.methods.transfer(to, amount), transactionOptions, SuperproToken.addressHttps);
            return store_1.default.web3Https.eth.getTransaction(receipt.transactionHash);
        });
    }
    /**
     * Approve tokens for specific address
     * @param address - address for approval
     * @param amount - number of tokens to be approved
     * @param transactionOptions - object what contains alternative action account or gas limit (optional)
     */
    static approve(address, amount, transactionOptions, checkTxBeforeSend = false) {
        return __awaiter(this, void 0, void 0, function* () {
            const contract = this.checkInit();
            (0, helper_1.checkIfActionAccountInitialized)(transactionOptions);
            if (checkTxBeforeSend) {
                yield TxManager_1.default.dryRun(contract.methods.approve(address, amount), transactionOptions);
            }
            yield TxManager_1.default.execute(contract.methods.approve(address, amount), transactionOptions, SuperproToken.addressHttps);
        });
    }
    static onTokenApprove(callback, owner, spender) {
        const contract = this.checkWssInit();
        const logger = this.logger.child({ method: 'onTokenApprove' });
        const subscription = contract.events.Approval();
        subscription.on('data', (event) => {
            const parsedEvent = (0, helper_1.cleanWeb3Data)(event.returnValues);
            if (owner && parsedEvent.owner != owner) {
                return;
            }
            if (spender && parsedEvent.spender != spender) {
                return;
            }
            callback(parsedEvent.owner, parsedEvent.spender, parsedEvent.value, {
                index: Number(event.blockNumber),
                hash: event.blockHash,
            });
        });
        subscription.on('error', (error) => {
            logger.warn(error);
        });
        return () => subscription.unsubscribe();
    }
    static onTokenTransfer(callback, from, to) {
        const contract = this.checkWssInit();
        const logger = this.logger.child({ method: 'onTokenTransfer' });
        const subscription = contract.events.Transfer();
        subscription.on('data', (event) => {
            const parsedEvent = (0, helper_1.cleanWeb3Data)(event.returnValues);
            if (from && parsedEvent.from != from) {
                return;
            }
            if (to && parsedEvent.to != to) {
                return;
            }
            callback(parsedEvent.from, parsedEvent.to, parsedEvent.value, {
                index: Number(event.blockNumber),
                hash: event.blockHash,
            });
        });
        subscription.on('error', (error) => {
            logger.warn(error);
        });
        return () => subscription.unsubscribe();
    }
}
SuperproToken.logger = logger_1.default.child({ className: 'SuperproToken' });
exports.default = SuperproToken;
