// @super-protocol/dto-js/build/enum/Cipher.enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cipher = void 0;
var Cipher;
(function (Cipher) {
    // Hybrid ciphers from external libraries
    Cipher["RSA_OAEP_AES_GCM"] = "{\"aesStandard\":\"AES-GCM\",\"rsaStandard\":\"RSA-OAEP\"}";
    // NodeJS native ciphers
    Cipher["AES_128_CBC_HMAC_SHA1"] = "aes-128-cbc-hmac-sha1";
    Cipher["AES_128_CBC_HMAC_SHA256"] = "aes-128-cbc-hmac-sha256";
    Cipher["AES_128_CCM"] = "aes-128-ccm";
    Cipher["AES_128_CFB"] = "aes-128-cfb";
    Cipher["AES_128_CFB1"] = "aes-128-cfb1";
    Cipher["AES_128_CFB8"] = "aes-128-cfb8";
    Cipher["AES_128_CTR"] = "aes-128-ctr";
    Cipher["AES_128_ECB"] = "aes-128-ecb";
    Cipher["AES_128_GCM"] = "aes-128-gcm";
    Cipher["AES_128_OFB"] = "aes-128-ofb";
    Cipher["AES_192_CBC"] = "aes-192-cbc";
    Cipher["AES_192_CCM"] = "aes-192-ccm";
    Cipher["AES_192_CFB"] = "aes-192-cfb";
    Cipher["AES_192_CFB1"] = "aes-192-cfb1";
    Cipher["AES_192_CFB8"] = "aes-192-cfb8";
    Cipher["AES_192_CTR"] = "aes-192-ctr";
    Cipher["AES_192_ECB"] = "aes-192-ecb";
    Cipher["AES_192_GCM"] = "aes-192-gcm";
    Cipher["AES_192_OFB"] = "aes-192-ofb";
    Cipher["AES_256_CBC"] = "aes-256-cbc";
    Cipher["AES_256_CBC_HMAC_SHA1"] = "aes-256-cbc-hmac-sha1";
    Cipher["AES_256_CBC_HMAC_SHA256"] = "aes-256-cbc-hmac-sha256";
    Cipher["AES_256_CCM"] = "aes-256-ccm";
    Cipher["AES_256_CFB"] = "aes-256-cfb";
    Cipher["AES_256_CFB1"] = "aes-256-cfb1";
    Cipher["AES_256_CFB8"] = "aes-256-cfb8";
    Cipher["AES_256_CTR"] = "aes-256-ctr";
    Cipher["AES_256_ECB"] = "aes-256-ecb";
    Cipher["AES_256_GCM"] = "aes-256-gcm";
    Cipher["AES_256_OFB"] = "aes-256-ofb";
    Cipher["AES128"] = "aes128";
    Cipher["AES192"] = "aes192";
    Cipher["AES256"] = "aes256";
    Cipher["ARIA_128_CBC"] = "aria-128-cbc";
    Cipher["ARIA_128_CCM"] = "aria-128-ccm";
    Cipher["ARIA_128_CFB"] = "aria-128-cfb";
    Cipher["ARIA_128_CFB1"] = "aria-128-cfb1";
    Cipher["ARIA_128_CFB8"] = "aria-128-cfb8";
    Cipher["ARIA_128_CTR"] = "aria-128-ctr";
    Cipher["ARIA_128_ECB"] = "aria-128-ecb";
    Cipher["ARIA_128_GCM"] = "aria-128-gcm";
    Cipher["ARIA_128_OFB"] = "aria-128-ofb";
    Cipher["ARIA_192_CBC"] = "aria-192-cbc";
    Cipher["ARIA_192_CCM"] = "aria-192-ccm";
    Cipher["ARIA_192_CFB"] = "aria-192-cfb";
    Cipher["ARIA_192_CFB1"] = "aria-192-cfb1";
    Cipher["ARIA_192_CFB8"] = "aria-192-cfb8";
    Cipher["ARIA_192_CTR"] = "aria-192-ctr";
    Cipher["ARIA_192_ECB"] = "aria-192-ecb";
    Cipher["ARIA_192_GCM"] = "aria-192-gcm";
    Cipher["ARIA_192_OFB"] = "aria-192-ofb";
    Cipher["ARIA_256_CBC"] = "aria-256-cbc";
    Cipher["ARIA_256_CCM"] = "aria-256-ccm";
    Cipher["ARIA_256_CFB"] = "aria-256-cfb";
    Cipher["ARIA_256_CFB1"] = "aria-256-cfb1";
    Cipher["ARIA_256_CFB8"] = "aria-256-cfb8";
    Cipher["ARIA_256_CTR"] = "aria-256-ctr";
    Cipher["ARIA_256_ECB"] = "aria-256-ecb";
    Cipher["ARIA_256_GCM"] = "aria-256-gcm";
    Cipher["ARIA_256_OFB"] = "aria-256-ofb";
    Cipher["ARIA128"] = "aria128";
    Cipher["ARIA192"] = "aria192";
    Cipher["ARIA256"] = "aria256";
    Cipher["BF"] = "bf";
    Cipher["BF_CBC"] = "bf-cbc";
    Cipher["BF_CFB"] = "bf-cfb";
    Cipher["BF_ECB"] = "bf-ecb";
    Cipher["BF_OFB"] = "bf-ofb";
    Cipher["BLOWFISH"] = "blowfish";
    Cipher["CAMELLIA_128_CBC"] = "camellia-128-cbc";
    Cipher["CAMELLIA_128_CFB"] = "camellia-128-cfb";
    Cipher["CAMELLIA_128_CFB1"] = "camellia-128-cfb1";
    Cipher["CAMELLIA_128_CFB8"] = "camellia-128-cfb8";
    Cipher["CAMELLIA_128_CTR"] = "camellia-128-ctr";
    Cipher["CAMELLIA_128_ECB"] = "camellia-128-ecb";
    Cipher["CAMELLIA_128_OFB"] = "camellia-128-ofb";
    Cipher["CAMELLIA_192_CBC"] = "camellia-192-cbc";
    Cipher["CAMELLIA_192_CFB"] = "camellia-192-cfb";
    Cipher["CAMELLIA_192_CFB1"] = "camellia-192-cfb1";
    Cipher["CAMELLIA_192_CFB8"] = "camellia-192-cfb8";
    Cipher["CAMELLIA_192_CTR"] = "camellia-192-ctr";
    Cipher["CAMELLIA_192_ECB"] = "camellia-192-ecb";
    Cipher["CAMELLIA_192_OFB"] = "camellia-192-ofb";
    Cipher["CAMELLIA_256_CBC"] = "camellia-256-cbc";
    Cipher["CAMELLIA_256_CFB"] = "camellia-256-cfb";
    Cipher["CAMELLIA_256_CFB1"] = "camellia-256-cfb1";
    Cipher["CAMELLIA_256_CFB8"] = "camellia-256-cfb8";
    Cipher["CAMELLIA_256_CTR"] = "camellia-256-ctr";
    Cipher["CAMELLIA_256_ECB"] = "camellia-256-ecb";
    Cipher["CAMELLIA_256_OFB"] = "camellia-256-ofb";
    Cipher["CAMELLIA128"] = "camellia128";
    Cipher["CAMELLIA192"] = "camellia192";
    Cipher["CAMELLIA256"] = "camellia256";
    Cipher["CAST"] = "cast";
    Cipher["CAST_CBC"] = "cast-cbc";
    Cipher["CAST5_CBC"] = "cast5-cbc";
    Cipher["CAST5_CFB"] = "cast5-cfb";
    Cipher["CAST5_ECB"] = "cast5-ecb";
    Cipher["CAST5_OFB"] = "cast5-ofb";
    Cipher["CHACHA20"] = "chacha20";
    Cipher["CHACHA20_POLY1305"] = "chacha20-poly1305";
    Cipher["DES"] = "des";
    Cipher["DES_CBC"] = "des-cbc";
    Cipher["DES_CFB"] = "des-cfb";
    Cipher["DES_CFB1"] = "des-cfb1";
    Cipher["DES_CFB8"] = "des-cfb8";
    Cipher["DES_ECB"] = "des-ecb";
    Cipher["DES_EDE"] = "des-ede";
    Cipher["DES_EDE_CBC"] = "des-ede-cbc";
    Cipher["DES_EDE_CFB"] = "des-ede-cfb";
    Cipher["DES_EDE_ECB"] = "des-ede-ecb";
    Cipher["DES_EDE_OFB"] = "des-ede-ofb";
    Cipher["DES_EDE3"] = "des-ede3";
    Cipher["DES_EDE3_CBC"] = "des-ede3-cbc";
    Cipher["DES_EDE3_CFB"] = "des-ede3-cfb";
    Cipher["DES_EDE3_CFB1"] = "des-ede3-cfb1";
    Cipher["DES_EDE3_CFB8"] = "des-ede3-cfb8";
    Cipher["DES_EDE3_ECB"] = "des-ede3-ecb";
    Cipher["DES_EDE3_OFB"] = "des-ede3-ofb";
    Cipher["DES_OFB"] = "des-ofb";
    Cipher["DES3"] = "des3";
    Cipher["DESX"] = "desx";
    Cipher["DESX_CBC"] = "desx-cbc";
    Cipher["ID_AES128_CCM"] = "id-aes128-CCM";
    Cipher["ID_AES128_GCM"] = "id-aes128-GCM";
    Cipher["ID_AES128_WRAP_PAD"] = "id-aes128-wrap-pad";
    Cipher["ID_AES192_CCM"] = "id-aes192-CCM";
    Cipher["ID_AES192_GCM"] = "id-aes192-GCM";
    Cipher["ID_AES192_WRAP_PAD"] = "id-aes192-wrap-pad";
    Cipher["ID_AES256_CCM"] = "id-aes256-CCM";
    Cipher["ID_AES256_GCM"] = "id-aes256-GCM";
    Cipher["ID_AES256_WRAP_PAD"] = "id-aes256-wrap-pad";
    Cipher["IDEA"] = "idea";
    Cipher["IDEA_CBC"] = "idea-cbc";
    Cipher["IDEA_CFB"] = "idea-cfb";
    Cipher["IDEA_ECB"] = "idea-ecb";
    Cipher["IDEA_OFB"] = "idea-ofb";
    Cipher["RC2"] = "rc2";
    Cipher["RC2_128"] = "rc2-128";
    Cipher["RC2_40"] = "rc2-40";
    Cipher["RC2_40_CBC"] = "rc2-40-cbc";
    Cipher["RC2_64"] = "rc2-64";
    Cipher["RC2_64_CBC"] = "rc2-64-cbc";
    Cipher["RC2_CBC"] = "rc2-cbc";
    Cipher["RC2_CFB"] = "rc2-cfb";
    Cipher["RC2_ECB"] = "rc2-ecb";
    Cipher["RC2_OFB"] = "rc2-ofb";
    Cipher["RC4"] = "rc4";
    Cipher["RC4_40"] = "rc4-40";
    Cipher["RC4_HMAC_MD5"] = "rc4-hmac-md5";
    Cipher["SEED"] = "seed";
    Cipher["SEED_CBC"] = "seed-cbc";
    Cipher["SEED_CFB"] = "seed-cfb";
    Cipher["SEED_ECB"] = "seed-ecb";
    Cipher["SEED_OFB"] = "seed-ofb";
    Cipher["SM4"] = "sm4";
    Cipher["SM4_CBC"] = "sm4-cbc";
    Cipher["SM4_CFB"] = "sm4-cfb";
    Cipher["SM4_CTR"] = "sm4-ctr";
    Cipher["SM4_ECB"] = "sm4-ecb";
    Cipher["SM4_OFB"] = "sm4-ofb";
    // TODO: add support for OCB ciphers
})(Cipher = exports.Cipher || (exports.Cipher = {}));
