// @super-protocol/dto-js/build/enum/EtlModel.enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EtlModelImageSubtype = exports.EtlModelStandardSubtype = exports.EtlModelType = void 0;
var EtlModelType;
(function (EtlModelType) {
    EtlModelType[EtlModelType["Standard"] = 0] = "Standard";
    EtlModelType[EtlModelType["Image"] = 1] = "Image";
    EtlModelType[EtlModelType["Custom"] = 2] = "Custom";
})(EtlModelType = exports.EtlModelType || (exports.EtlModelType = {}));
var EtlModelStandardSubtype;
(function (EtlModelStandardSubtype) {
    EtlModelStandardSubtype[EtlModelStandardSubtype["Number"] = 0] = "Number";
    EtlModelStandardSubtype[EtlModelStandardSubtype["String"] = 1] = "String";
    EtlModelStandardSubtype[EtlModelStandardSubtype["Boolean"] = 2] = "Boolean";
})(EtlModelStandardSubtype = exports.EtlModelStandardSubtype || (exports.EtlModelStandardSubtype = {}));
var EtlModelImageSubtype;
(function (EtlModelImageSubtype) {
    EtlModelImageSubtype[EtlModelImageSubtype["JPEG"] = 0] = "JPEG";
    EtlModelImageSubtype[EtlModelImageSubtype["PNG"] = 1] = "PNG";
    EtlModelImageSubtype[EtlModelImageSubtype["BMP"] = 2] = "BMP";
    EtlModelImageSubtype[EtlModelImageSubtype["TIFF"] = 3] = "TIFF";
})(EtlModelImageSubtype = exports.EtlModelImageSubtype || (exports.EtlModelImageSubtype = {}));
