// @super-protocol/sdk-js/build/tee/statuses.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QuoteValidationStatuses = exports.TCBStatuses = exports.QEIdentityStatuses = void 0;
var QEIdentityStatuses;
(function (QEIdentityStatuses) {
    QEIdentityStatuses["UpToDate"] = "UpToDate";
    QEIdentityStatuses["OutOfDate"] = "OutOfDate";
    QEIdentityStatuses["Revoked"] = "Revoked";
})(QEIdentityStatuses = exports.QEIdentityStatuses || (exports.QEIdentityStatuses = {}));
var TCBStatuses;
(function (TCBStatuses) {
    TCBStatuses["UpToDate"] = "UpToDate";
    TCBStatuses["OutOfDate"] = "OutOfDate";
    TCBStatuses["Revoked"] = "Revoked";
    TCBStatuses["ConfigurationNeeded"] = "ConfigurationNeeded";
    TCBStatuses["ConfigurationAndSWHardeningNeeded"] = "ConfigurationAndSWHardeningNeeded";
    TCBStatuses["SWHardeningNeeded"] = "SWHardeningNeeded";
    TCBStatuses["OutOfDateConfigurationNeeded"] = "OutOfDateConfigurationNeeded";
})(TCBStatuses = exports.TCBStatuses || (exports.TCBStatuses = {}));
var QuoteValidationStatuses;
(function (QuoteValidationStatuses) {
    QuoteValidationStatuses["UpToDate"] = "UpToDate";
    QuoteValidationStatuses["ConfigurationNeeded"] = "ConfigurationNeeded";
    QuoteValidationStatuses["SecurityPatchNeeded"] = "SecurityPatchNeeded";
    QuoteValidationStatuses["SoftwareUpdateNeeded"] = "SoftwareUpdateNeeded";
    QuoteValidationStatuses["Error"] = "Error";
})(QuoteValidationStatuses = exports.QuoteValidationStatuses || (exports.QuoteValidationStatuses = {}));
