// @super-protocol/sdk-js/build/utils/Monitoring.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Monitoring = void 0;
const logger_1 = __importDefault(require("../logger"));
class Monitoring {
    constructor() {
        this.logger = logger_1.default.child({ className: 'Monitoring' });
        this.contractMethodCalls = new Map();
    }
    static getInstance() {
        if (!Monitoring.instance) {
            Monitoring.instance = new Monitoring();
        }
        return Monitoring.instance;
    }
    initializeLogging() {
        const checkInterval = process.env.PRINT_CONTRACT_CALLS_INTERVAL || 300000;
        this.shutdownLogging();
        const startTs = Date.now();
        this.interval = setInterval(() => {
            this.contractMethodCalls.forEach((value, key) => this.logger.debug({
                methodName: key,
                calledTimes: value,
            }));
            const totalCalls = Array.from(this.contractMethodCalls.values()).reduce((acc, curr) => acc + curr, 0);
            const timeSpend = Math.floor((Date.now() - startTs) / (60 * 1000));
            this.logger.debug({
                timeSpend: timeSpend + ' min',
                totalCalls,
            }, 'Contract methods calls');
        }, +checkInterval);
    }
    shutdownLogging() {
        if (this.interval) {
            clearInterval(this.interval);
            this.interval = undefined;
        }
    }
    incrementCall(methodName) {
        const prevValue = this.contractMethodCalls.get(methodName) || 0;
        this.contractMethodCalls.set(methodName, prevValue + 1);
    }
}
exports.Monitoring = Monitoring;
