// @super-protocol/sdk-js/build/utils/NonceTracker.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const logger_1 = __importDefault(require("../logger"));
class NonceTracker {
    constructor(web3, address) {
        this.web3 = web3;
        this.address = address;
        this.countOfPendingTransactions = 0;
        this.logger = logger_1.default.child({ className: 'NonceTracker', address });
        this.logger.trace('Created NonceTracker');
    }
    initAccount() {
        return __awaiter(this, void 0, void 0, function* () {
            this.txCount = yield this.web3.eth.getTransactionCount(this.address);
            this.logger.trace(`Initialized ${this.address} account with nonce: ${this.txCount}`);
        });
    }
    getNonce() {
        if (this.txCount === undefined)
            throw Error(`NonceTracker for address ${this.address} is not initialized`);
        this.logger.trace(`Get nonce: ${this.txCount}`);
        return this.txCount;
    }
    consumeNonce() {
        if (this.txCount === undefined)
            throw Error(`NonceTracker for address ${this.address} is not initialized`);
        this.logger.trace(`Consume nonce: ${this.txCount + BigInt(1)}`);
        return this.txCount++;
    }
    onTransactionStartPublishing() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.transactionsOnHold) {
                yield this.waitForPendingTransactions();
            }
            this.countOfPendingTransactions++;
        });
    }
    onTransactionPublished() {
        this.countOfPendingTransactions--;
        if (this.countOfPendingTransactions === 0) {
            void this.sendHoldTransactions();
        }
    }
    onTransactionError() {
        return __awaiter(this, void 0, void 0, function* () {
            this.countOfPendingTransactions--;
            if (!this.transactionsOnHold)
                this.transactionsOnHold = [];
            if (this.countOfPendingTransactions === 0) {
                yield this.sendHoldTransactions();
            }
            else {
                yield this.waitForPendingTransactions();
            }
        });
    }
    waitForPendingTransactions() {
        return new Promise((resolve) => {
            if (!this.transactionsOnHold)
                return resolve();
            this.transactionsOnHold.push(() => {
                resolve();
            });
        });
    }
    sendHoldTransactions() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.transactionsOnHold)
                return;
            yield this.initAccount();
            this.transactionsOnHold.forEach((callback) => callback());
            this.transactionsOnHold = undefined;
        });
    }
}
exports.default = NonceTracker;
