// @super-protocol/sdk-js/build/utils/TxManager.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Web3TransactionRevertedByEvmError = void 0;
const NonceTracker_1 = __importDefault(require("./NonceTracker"));
const logger_1 = __importDefault(require("../logger"));
const store_1 = __importDefault(require("../store"));
const helper_1 = require("./helper");
const Superpro_1 = __importDefault(require("../staticModels/Superpro"));
const constants_1 = require("../constants");
const lodash_1 = __importDefault(require("lodash"));
const bottleneck_1 = __importDefault(require("bottleneck"));
class Web3TransactionError extends Error {
    constructor(originalError, message) {
        super(message);
        this.name = 'Web3TransactionError';
        this.originalError = originalError;
    }
}
class Web3TransactionRevertedByEvmError extends Web3TransactionError {
    constructor(originalError, message) {
        super(originalError, message);
        this.name = 'Web3TransactionRevertedByEvmError';
    }
}
exports.Web3TransactionRevertedByEvmError = Web3TransactionRevertedByEvmError;
class TxManager {
    static init(web3) {
        this.web3 = web3;
    }
    static checkIfInitialized() {
        if (!this.web3) {
            throw Error('TxManager should be initialized before using.');
        }
    }
    static initAccount(address) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.nonceTrackers[address])
                return;
            this.nonceTrackers[address] = new NonceTracker_1.default(this.web3, address);
            yield this.nonceTrackers[address].initAccount();
        });
    }
    static execute(transaction, transactionOptions, to = Superpro_1.default.address) {
        const txData = {
            to,
            data: transaction.encodeABI(),
        };
        return TxManager.publishTransaction(txData, transactionOptions, transaction);
    }
    static publishTransaction(txData, transactionOptions, transactionCall) {
        return __awaiter(this, void 0, void 0, function* () {
            this.checkIfInitialized();
            (0, helper_1.checkIfActionAccountInitialized)(transactionOptions);
            const txOptions = yield (0, helper_1.createTransactionOptions)(Object.assign({}, transactionOptions));
            if (!txOptions.from) {
                throw Error('From account is undefined. You should pass it to transactionOptions or init action account.');
            }
            const publishTxOptions = Object.assign(Object.assign({}, txOptions), { web3: (transactionOptions === null || transactionOptions === void 0 ? void 0 : transactionOptions.web3) || this.web3, useExternalTxManager: (0, helper_1.checkForUsingExternalTxManager)(transactionOptions) });
            if (!this.queues[publishTxOptions.from]) {
                this.queues[publishTxOptions.from] = new bottleneck_1.default({
                    maxConcurrent: store_1.default.txConcurrency,
                    minTime: store_1.default.txIntervalMs,
                });
            }
            return this.queues[publishTxOptions.from].schedule(() => TxManager._publishTransaction(txData, publishTxOptions, transactionCall));
        });
    }
    static dryRun(transaction, transactionOptions) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const from = (_a = transactionOptions === null || transactionOptions === void 0 ? void 0 : transactionOptions.from) !== null && _a !== void 0 ? _a : store_1.default.actionAccount;
            try {
                return yield transaction.call({ from });
            }
            catch (e) {
                e.txErrorMsg = e.data.message || 'Error text is undefined';
                throw e;
            }
        });
    }
    static _publishTransaction(txData, transactionOptions, transactionCall) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const { from, gas, gasPrice, gasPriceMultiplier, web3 } = transactionOptions;
            txData = Object.assign(Object.assign({}, txData), { from,
                gas,
                gasPrice,
                gasPriceMultiplier });
            if (transactionCall) {
                let estimatedGas;
                try {
                    estimatedGas = yield transactionCall.estimateGas(txData);
                }
                catch (e) {
                    TxManager.logger.debug({ error: e }, 'Fail to calculate estimated gas');
                    estimatedGas = constants_1.defaultGasLimit;
                }
                txData.gas = (0, helper_1.multiplyBigIntByNumber)(estimatedGas, store_1.default.gasLimitMultiplier);
                // defaultGasLimit is max gas limit
                txData.gas = txData.gas < constants_1.defaultGasLimit ? txData.gas : constants_1.defaultGasLimit;
                if (transactionOptions.gas) {
                    if (transactionOptions.gas < estimatedGas) {
                        TxManager.logger.warn({
                            estimated: estimatedGas,
                            specified: transactionOptions.gas,
                        }, 'Fail to calculate estimated gas');
                    }
                    txData.gas = transactionOptions.gas;
                }
                txData.gasPrice = (0, helper_1.multiplyBigIntByNumber)(txData.gasPrice, store_1.default.gasPriceMultiplier);
            }
            let nonceTracker;
            // TODO: Consider a better way to organize different strategies for publishing transactions.
            if (!transactionOptions.useExternalTxManager && this.nonceTrackers[transactionOptions.from]) {
                nonceTracker = this.nonceTrackers[transactionOptions.from];
                yield nonceTracker.onTransactionStartPublishing();
                txData.nonce = nonceTracker.consumeNonce();
            }
            const signingKey = store_1.default.keys[transactionOptions.from];
            try {
                let transactionResultData;
                if (signingKey) {
                    const signed = yield web3.eth.accounts.signTransaction(txData, signingKey);
                    if (!signed.rawTransaction) {
                        throw new Error('Failed to sign transaction');
                    }
                    TxManager.logger.debug({
                        txHash: signed.transactionHash,
                        txData: lodash_1.default.omit(txData, ['data']),
                    }, 'Publishing signed transaction');
                    transactionResultData = yield web3.eth.sendSignedTransaction(signed.rawTransaction);
                    TxManager.logger.debug({
                        txHash: signed.transactionHash,
                        txBlockNumber: transactionResultData.blockNumber,
                        txGasUsed: transactionResultData.gasUsed,
                    }, 'Transaction result');
                }
                else {
                    TxManager.logger.debug({
                        txData: lodash_1.default.omit(txData, ['data']),
                    }, 'Publishing unsigned transaction');
                    transactionResultData = yield web3.eth.sendTransaction(txData);
                }
                if (nonceTracker)
                    nonceTracker.onTransactionPublished();
                return transactionResultData;
            }
            catch (e) {
                const message = 'Error during transaction execution';
                TxManager.logger.error(e, message);
                if (nonceTracker)
                    yield nonceTracker.onTransactionError();
                if ((_a = e.message) === null || _a === void 0 ? void 0 : _a.includes('Transaction has been reverted by the EVM')) {
                    throw new Web3TransactionRevertedByEvmError(e, message);
                }
                else {
                    throw new Web3TransactionError(e, message);
                }
            }
        });
    }
}
TxManager.logger = logger_1.default.child({ className: 'TxManager' });
TxManager.nonceTrackers = {};
TxManager.queues = {};
exports.default = TxManager;
