// @super-protocol/addons-tee/dist/sgx-native-module/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PkiServiceError = exports.QuoteSecurityValidationError = exports.QuoteValidationError = exports.AttestationDcapQuoteVerifyEnclaveError = void 0;
const consts_1 = require("./consts");
class AttestationDcapQuoteVerifyEnclaveError extends Error {
    constructor(msg) {
        super(msg);
        this.name = AttestationDcapQuoteVerifyEnclaveError.name;
    }
}
exports.AttestationDcapQuoteVerifyEnclaveError = AttestationDcapQuoteVerifyEnclaveError;
class QuoteValidationError extends AttestationDcapQuoteVerifyEnclaveError {
    constructor(verifyResult) {
        super();
        this.verifyResult = verifyResult;
        this.criticalError = false;
        this.verifyResult = verifyResult;
        this.criticalError = false;
        switch (verifyResult) {
            case 0xa001:
                this.message = `The SGX platform firmware and SW are at the latest security patching level but there are
                platform hardware configurations may expose the enclave to vulnerabilities.`;
                break;
            case 0xa002:
            case 0xa003:
            case 0xa004:
                this.message = `The SGX platform firmware and SW are not at the latest security patching level. The
                platform needs to be patched with firmware and/or software patches.`;
                break;
            case 0xa007:
            case 0xa008:
                this.message = `The SGX platform firmware and SW are at the latest security patching level but there
                are certain vulnerabilities that can only be mitigated with software mitigations implemented by the enclave.`;
                break;
            default:
                this.criticalError = true;
                this.message = `Quote verification failed. Verification result: 0x${Number(verifyResult).toString(16)}`;
                break;
        }
    }
}
exports.QuoteValidationError = QuoteValidationError;
class QuoteSecurityValidationError extends AttestationDcapQuoteVerifyEnclaveError {
    constructor(verifyResult) {
        super();
        this.verifyResult = verifyResult;
        this.verifyResult = verifyResult;
        this.message = "Invalid CPU settings are being used.";
        if (verifyResult.smtEnabled === consts_1.PckFlag.PCK_FLAG_TRUE) {
            this.message = " HyperThreading (SMT) is enabled. Please disable it in BIOS.";
        }
    }
}
exports.QuoteSecurityValidationError = QuoteSecurityValidationError;
class PkiServiceError extends Error {
    constructor(msg) {
        super(msg);
        this.name = PkiServiceError.name;
    }
}
exports.PkiServiceError = PkiServiceError;
