// @super-protocol/addons-tee/dist/sgx-native-module/sgx-tests.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const SgxNative = __importStar(require("../sgx-native-module"));
const assert_1 = __importDefault(require("assert"));
const dcapVerifyService = new SgxNative.DcapQuoteVerifyService();
const pkiService = new SgxNative.PkiService();
const enclaveService = new SgxNative.EnclaveService();
async function GenCertWithQouteTest() {
    const res = await pkiService.generateTlsCertificate({
        days: 14,
        dnsNames: ["superprotocol.com", "sp.com"],
        ips: ["1.2.3.4", "2001:0db8:85a3:0000:0000:8a2e:0370:7334"],
        format: SgxNative.CertificateFormat.DER,
        keyType: SgxNative.CertificateKeyType.ECP,
        ecpCurve: SgxNative.ECPCurve.SECP384R1,
        subject: {
            commonName: "localhost",
        },
        serialNumber: "12345",
        withQuote: true,
    });
    try {
        const quote = await dcapVerifyService.extractQuoteFromCert(res.cert);
        await dcapVerifyService.validateQuoteVerifierEnclave(quote);
    }
    catch (error) {
        if (error instanceof SgxNative.QuoteValidationError) {
            assert_1.default.equal(error.criticalError, false);
        }
        else {
            (0, assert_1.default)(false);
        }
    }
    console.log("GenCertWithQouteTest: Success");
}
async function GetKeyTest() {
    const key = await enclaveService.getSecretKey(SgxNative.KeyType.SGX_KEYSELECT_SEAL, SgxNative.KeyPolicy.SGX_KEYPOLICY_MRSIGNER);
    const key2 = await enclaveService.getSecretKey(SgxNative.KeyType.SGX_KEYSELECT_SEAL, SgxNative.KeyPolicy.SGX_KEYPOLICY_MRSIGNER, key.request);
    const key3 = await enclaveService.getSecretKey(SgxNative.KeyType.SGX_KEYSELECT_SEAL, SgxNative.KeyPolicy.SGX_KEYPOLICY_MRSIGNER);
    (0, assert_1.default)(Buffer.compare(key.key, key2.key) === 0);
    (0, assert_1.default)(Buffer.compare(key.key, key3.key) !== 0);
    await assert_1.default.rejects(async () => {
        await enclaveService.getSecretKey(SgxNative.KeyType.SGX_KEYSELECT_SEAL, SgxNative.KeyPolicy.SGX_KEYPOLICY_MRENCLAVE, key.request);
    }, {
        name: "TypeError",
        message: "The previous request parameters do not match the request name or policy",
    });
    const outDated = await enclaveService.isOutdatedRequest(key.request);
    (0, assert_1.default)(outDated === false);
    const baseCpuSvn = Buffer.alloc(16);
    key.request.copy(baseCpuSvn, 0, 8, 24);
    let downgradeIndex = -1;
    let upgradeIndex = -1;
    for (let i = 0; i < 16; i++) {
        if (downgradeIndex < 0 && baseCpuSvn[i] < 0xff)
            downgradeIndex = i;
        if (upgradeIndex < 0 && baseCpuSvn[i] > 0)
            upgradeIndex = i;
        if (upgradeIndex >= 0 && downgradeIndex >= 0)
            break;
    }
    const modifiedRequestDowngrade = Buffer.alloc(key.request.length);
    key.request.copy(modifiedRequestDowngrade);
    const modifiedRequestUpgrade = Buffer.alloc(key.request.length);
    key.request.copy(modifiedRequestUpgrade);
    modifiedRequestDowngrade[8 + downgradeIndex] = modifiedRequestDowngrade[8 + downgradeIndex] - 1;
    modifiedRequestUpgrade[8 + upgradeIndex] = modifiedRequestUpgrade[8 + upgradeIndex] + 1;
    const outDated2 = await enclaveService.isOutdatedRequest(modifiedRequestDowngrade);
    (0, assert_1.default)(outDated2 === true);
    await assert_1.default.rejects(async () => {
        await enclaveService.isOutdatedRequest(modifiedRequestUpgrade);
    }, {
        name: "TypeError",
        message: /Failed to obtain the key with the passed request*/,
    });
    await assert_1.default.rejects(async () => {
        await enclaveService.getSecretKey(SgxNative.KeyType.SGX_KEYSELECT_SEAL, SgxNative.KeyPolicy.SGX_KEYPOLICY_MRSIGNER, modifiedRequestUpgrade);
    }, {
        name: "TypeError",
        message: /Failed to obtain the key with the passed request*/,
    });
    console.log("GetKeyTest: Success");
}
(async () => {
    await GenCertWithQouteTest();
    await GetKeyTest();
})();
