// @super-protocol/sdk-js/build/models/EtlModel.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EtlModel = void 0;
const protobuf = __importStar(require("protobufjs"));
const logger_1 = __importDefault(require("../logger"));
const ModelPackager_1 = require("../staticModels/ModelPackager");
const ResourceContentType_enum_1 = require("@super-protocol/dto-js/build/enum/ResourceContentType.enum");
const resourceLoaders_1 = require("../utils/resourceLoaders");
class EtlModel {
    constructor(etlModel) {
        this.etlModel = etlModel;
    }
    /**
     * Create instance of EtlModel from Buffer
     *
     * @param data - packed EtlModel
     * @returns instance of EtlModel
     */
    static unpack(data) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const etlModel = yield ModelPackager_1.ModelPackager.unpack(data);
                return new EtlModel(etlModel);
            }
            catch (error) {
                EtlModel.logger.error('Unable to unpack EtlModel');
                throw error;
            }
        });
    }
    /**
     * Packing EltModel
     *
     * @returns EltModel in binary format
     */
    pack() {
        return ModelPackager_1.ModelPackager.pack(this.etlModel);
    }
    downloadMetadata() {
        return __awaiter(this, void 0, void 0, function* () {
            const { metadata } = this.etlModel;
            if (!(metadata === null || metadata === void 0 ? void 0 : metadata.resource)) {
                throw new Error('Resource is not present int he EtlModel');
            }
            const ResourceLoader = (0, resourceLoaders_1.getResourceLoader)(metadata.resource.type);
            try {
                const bytes = yield new ResourceLoader().download(metadata.resource);
                if (bytes.byteLength === 0) {
                    throw new Error('Resource is empty');
                }
                const stringified = bytes.toString();
                switch (metadata.resourceContentType) {
                    case ResourceContentType_enum_1.ResourceContentType.JSON:
                        return this.parseJsonResource(stringified);
                    case ResourceContentType_enum_1.ResourceContentType.PROTOBUF:
                        return this.parseProtobufResource(stringified);
                    default:
                        throw new Error(`Resource content type ${metadata.resourceContentType} is not supported`);
                }
            }
            catch (error) {
                throw new Error(`Error during download and parsing resource: ${error.message}`);
            }
        });
    }
    getType() {
        return this.etlModel.type;
    }
    setType(type) {
        this.etlModel.type = type;
    }
    getSubtype() {
        return this.etlModel.subtype;
    }
    setSubtype(subtype) {
        this.etlModel.subtype = subtype;
    }
    getMetadata() {
        return this.etlModel.metadata;
    }
    setMetadata(metadata) {
        this.etlModel.metadata = metadata;
    }
    parseJsonResource(data) {
        try {
            return JSON.parse(data);
        }
        catch (error) {
            throw new Error('JSON data in incorrect');
        }
    }
    parseProtobufResource(data) {
        try {
            return protobuf.parse(data).root.toJSON().nested;
        }
        catch (error) {
            throw new Error('Protobuf message is incorrect');
        }
    }
}
exports.EtlModel = EtlModel;
EtlModel.logger = logger_1.default.child({ name: EtlModel.name });
