// @super-protocol/sdk-js/build/models/Offer.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const logger_1 = __importDefault(require("../logger"));
const helper_1 = require("../utils/helper");
const connectors_1 = require("../connectors");
const types_1 = require("../types");
const utils_1 = require("ethers/lib/utils");
const TeeOffers_1 = __importDefault(require("../staticModels/TeeOffers"));
const TxManager_1 = __importDefault(require("../utils/TxManager"));
const helpers_1 = require("../utils/helpers");
const constants_1 = require("../constants");
class Offer {
    constructor(offerId) {
        this.id = offerId;
        if (!Offer.contract) {
            Offer.contract = connectors_1.BlockchainConnector.getInstance().getContract();
        }
        this.logger = logger_1.default.child({
            className: 'Offer',
            offerId: this.id,
        });
    }
    /**
     * Function for fetching offer status from blockchain
     */
    isEnabled() {
        return Offer.contract.methods.isOfferEnabled(this.id).call();
    }
    /**
     * Updates name in offer info
     * @param name - new name
     * @param transactionOptions - object what contains alternative action account or gas limit (optional)
     */
    setName(name, transactionOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            (0, helper_1.checkIfActionAccountInitialized)(transactionOptions);
            yield TxManager_1.default.execute(Offer.contract.methods.setOfferName(this.id, name), transactionOptions);
            if (this.offerInfo)
                this.offerInfo.name = name;
        });
    }
    /**
     * Updates description in offer info
     * @param description - new description
     * @param transactionOptions - object what contains alternative action account or gas limit (optional)
     */
    setDescription(description, transactionOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            (0, helper_1.checkIfActionAccountInitialized)(transactionOptions);
            yield TxManager_1.default.execute(Offer.contract.methods.setOfferDescription(this.id, description), transactionOptions);
            if (this.offerInfo)
                this.offerInfo.description = description;
        });
    }
    /**
     * Updates offer info
     * @param newInfo - new offer info
     * @param transactionOptions - object what contains alternative action account or gas limit (optional)
     */
    setInfo(newInfo, transactionOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            (0, helper_1.checkIfActionAccountInitialized)(transactionOptions);
            yield TxManager_1.default.execute(Offer.contract.methods.setValueOfferInfo(this.id, newInfo), transactionOptions);
            if (this.offerInfo)
                this.offerInfo = newInfo;
        });
    }
    /**
     * Function for fetching offer info from blockchain
     */
    getInfo() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!(yield this.checkIfOfferExistsWithInterval())) {
                throw Error(`Offer ${this.id} does not exist`);
            }
            const { info } = yield Offer.contract.methods.getValueOffer(this.id).call();
            this.offerInfo = (0, helper_1.cleanWeb3Data)(info);
            return this.offerInfo;
        });
    }
    /**
     * Function for fetching offer provider from blockchain (works for TEE and Value offers)
     */
    getProvider() {
        return __awaiter(this, void 0, void 0, function* () {
            this.provider = yield Offer.contract.methods.getOfferProviderAuthority(this.id).call();
            return this.provider;
        });
    }
    /**
     * Fetch offer type from blockchain (works for TEE and Value offers)
     */
    getOfferType() {
        return __awaiter(this, void 0, void 0, function* () {
            this.type = yield Offer.contract.methods.getOfferType(this.id).call();
            return this.type.toString();
        });
    }
    /**
     * Function for fetching TEE offer provider authority account from blockchain
     */
    getProviderAuthority() {
        return __awaiter(this, void 0, void 0, function* () {
            this.providerAuthority = yield Offer.contract.methods.getOfferProviderAuthority(this.id).call();
            return this.providerAuthority;
        });
    }
    /**
     * Fetch new Origins (createdDate, createdBy, modifiedDate and modifiedBy)
     */
    getOrigins() {
        return __awaiter(this, void 0, void 0, function* () {
            const origins = yield Offer.contract.methods
                .getOfferOrigins(this.id)
                .call()
                .then((origins) => (0, helper_1.cleanWeb3Data)(origins));
            // Convert blockchain time seconds to js time milliseconds
            origins.createdDate = Number(origins.createdDate) * 1000;
            origins.modifiedDate = Number(origins.modifiedDate) * 1000;
            return (this.origins = origins);
        });
    }
    /**
     * Function for fetching offer hold deposit
     */
    getMinDeposit(slotId) {
        return __awaiter(this, void 0, void 0, function* () {
            this.minDeposit = yield Offer.contract.methods
                .getOfferMinDeposit(this.id, slotId, '0', [], [])
                .call()
                .then((price) => (0, helper_1.convertBigIntToString)(price));
            return this.minDeposit;
        });
    }
    /**
     * Function for fetching cheapest value offer from blockchain
     */
    getCheapestPrice() {
        return Offer.contract.methods
            .getCheapestValueOffersPrice(this.id)
            .call()
            .then((price) => (0, helper_1.convertBigIntToString)(price));
    }
    isOfferExists() {
        return Offer.contract.methods.isOfferExists(this.id).call();
    }
    checkIfOfferExistsWithInterval() {
        return __awaiter(this, void 0, void 0, function* () {
            const offerExists = yield (0, helpers_1.tryWithInterval)({
                handler: () => this.isOfferExists(),
                checkResult: (exists) => {
                    if (!exists)
                        this.logger.debug(`Offer ${this.id} exists: ${exists}`);
                    return { isResultOk: exists };
                },
                retryInterval: constants_1.BLOCKCHAIN_CALL_RETRY_INTERVAL,
                retryMax: constants_1.BLOCKCHAIN_CALL_RETRY_ATTEMPTS,
            });
            return offerExists;
        });
    }
    /**
     * Function for fetching whether offer slot exists or not
     * @param slotId - Slot ID
     */
    isSlotExists(slotId) {
        return Offer.contract.methods.isValueOfferSlotExists(this.id, slotId).call();
    }
    /**
     * Function for fetching offer slot by id
     * @param slotId - Slot ID
     */
    getSlotById(slotId) {
        return __awaiter(this, void 0, void 0, function* () {
            const slot = yield Offer.contract.methods
                .getValueOfferSlotById(this.id, slotId)
                .call();
            const cpuDenominator = yield TeeOffers_1.default.getDenominator();
            return (0, helper_1.formatOfferSlot)(slot, cpuDenominator);
        });
    }
    /**
     * @returns this TEE offer slots count
     */
    getSlotsCount() {
        return __awaiter(this, void 0, void 0, function* () {
            return Number(yield Offer.contract.methods.getValueOfferSlotsCount(this.id).call());
        });
    }
    /**
     * Function for fetching  offer slots info from blockchain
     * @param begin - The first element of range.
     * @param end - One past the final element in the range.
     * @returns {Promise<ValueOfferSlot[]>}
     */
    getSlots(begin = 0, end = 999999) {
        return __awaiter(this, void 0, void 0, function* () {
            const slotsCount = Number(yield Offer.contract.methods.getValueOfferSlotsCount(this.id).call());
            if (slotsCount === 0) {
                return [];
            }
            const slots = yield Offer.contract.methods
                .getValueOfferSlots(this.id, begin, end)
                .call()
                .then((slots) => slots.map((slot) => (0, helper_1.transformComplexObject)(slot)));
            const cpuDenominator = yield TeeOffers_1.default.getDenominator();
            const slotsResult = slots.map((slot) => (0, helper_1.formatOfferSlot)(slot, cpuDenominator));
            return slotsResult;
        });
    }
    /**
     * Function for add slot usage to the value offer
     * @param slotInfo - slot info
     * @param optionInfo - option info
     * @param slotUsage - slot usage info
     * @param transactionOptions - object what contains alternative action account or gas limit (optional)
     */
    addSlot(slotInfo, optionInfo, slotUsage, externalId = 'default', transactionOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            (0, helper_1.checkIfActionAccountInitialized)(transactionOptions);
            slotInfo = (0, helper_1.packSlotInfo)(slotInfo, yield TeeOffers_1.default.getDenominator());
            const formattedExternalId = (0, utils_1.formatBytes32String)(externalId);
            const transactionCall = Offer.contract.methods.addValueOfferSlot(this.id, formattedExternalId, slotInfo, optionInfo, slotUsage);
            yield TxManager_1.default.execute(transactionCall, transactionOptions);
        });
    }
    /**
     * Function for update slot usage of value offer
     * @param slotInfo - new slot info
     * @param optionInfo - new option info
     * @param slotUsage - new slot usage info
     * @param transactionOptions - object what contains alternative action account or gas limit (optional)
     */
    updateSlot(slotId, newSlotInfo, newOptionInfo, newUsage, transactionOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            (0, helper_1.checkIfActionAccountInitialized)(transactionOptions);
            newSlotInfo = (0, helper_1.packSlotInfo)(newSlotInfo, yield TeeOffers_1.default.getDenominator());
            yield TxManager_1.default.execute(Offer.contract.methods.updateValueOfferSlot(this.id, slotId, newSlotInfo, newOptionInfo, newUsage), transactionOptions);
        });
    }
    /**
     * Function for delete slot usage from offer
     * @param slotId - Slot ID
     * @param transactionOptions - object what contains alternative action account or gas limit (optional)
     */
    deleteSlot(slotId, transactionOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            (0, helper_1.checkIfActionAccountInitialized)(transactionOptions);
            yield TxManager_1.default.execute(Offer.contract.methods.deleteValueOfferSlot(this.id, slotId), transactionOptions);
        });
    }
    /**
     * Function for disabling offer
     * @param transactionOptions - object what contains alternative action account or gas limit (optional)
     */
    disable(transactionOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            (0, helper_1.checkIfActionAccountInitialized)(transactionOptions);
            yield TxManager_1.default.execute(Offer.contract.methods.disableOffer(this.id), transactionOptions);
        });
    }
    /**
     * Function for enabling offer
     * @param transactionOptions - object what contains alternative action account or gas limit (optional)
     */
    enable(transactionOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            (0, helper_1.checkIfActionAccountInitialized)(transactionOptions);
            yield TxManager_1.default.execute(Offer.contract.methods.enableOffer(this.id), transactionOptions);
        });
    }
    /**
     * Checks if passed offer match restrictions in this offer
     * @param offerId - id of offer what needs to be checked
     */
    isRestrictionsPermitThatOffer(offerId) {
        return Offer.contract.methods.isOfferRestrictionsPermitOtherOffer(this.id, offerId).call();
    }
    /**
     * Checks if this offer contains restrictions of a certain type
     * @param type - type of offer which needs to be checked
     */
    isRestrictedByOfferType(type) {
        return Offer.contract.methods.isOfferRestrictedByOfferType(this.id, type).call();
    }
}
__decorate([
    (0, helper_1.incrementMethodCall)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], Offer.prototype, "isEnabled", null);
__decorate([
    (0, helper_1.incrementMethodCall)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], Offer.prototype, "getInfo", null);
__decorate([
    (0, helper_1.incrementMethodCall)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], Offer.prototype, "getProvider", null);
__decorate([
    (0, helper_1.incrementMethodCall)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], Offer.prototype, "getOfferType", null);
__decorate([
    (0, helper_1.incrementMethodCall)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], Offer.prototype, "getProviderAuthority", null);
__decorate([
    (0, helper_1.incrementMethodCall)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], Offer.prototype, "getOrigins", null);
__decorate([
    (0, helper_1.incrementMethodCall)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", Promise)
], Offer.prototype, "getMinDeposit", null);
__decorate([
    (0, helper_1.incrementMethodCall)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], Offer.prototype, "getCheapestPrice", null);
__decorate([
    (0, helper_1.incrementMethodCall)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], Offer.prototype, "isOfferExists", null);
__decorate([
    (0, helper_1.incrementMethodCall)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object, Object, Object, Object]),
    __metadata("design:returntype", Promise)
], Offer.prototype, "addSlot", null);
__decorate([
    (0, helper_1.incrementMethodCall)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object, Object, Object, Object]),
    __metadata("design:returntype", Promise)
], Offer.prototype, "updateSlot", null);
__decorate([
    (0, helper_1.incrementMethodCall)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object]),
    __metadata("design:returntype", Promise)
], Offer.prototype, "deleteSlot", null);
__decorate([
    (0, helper_1.incrementMethodCall)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", Promise)
], Offer.prototype, "isRestrictionsPermitThatOffer", null);
__decorate([
    (0, helper_1.incrementMethodCall)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", Promise)
], Offer.prototype, "isRestrictedByOfferType", null);
exports.default = Offer;
