// @super-protocol/sdk-js/build/models/Provider.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const helper_1 = require("../utils/helper");
const connectors_1 = require("../connectors");
const TxManager_1 = __importDefault(require("../utils/TxManager"));
const Consensus_1 = __importDefault(require("../staticModels/Consensus"));
class Provider {
    constructor(providerId) {
        this.providerId = providerId;
        if (!Provider.contract) {
            Provider.contract = connectors_1.BlockchainConnector.getInstance().getContract();
        }
    }
    modify(providerInfo, transactionOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            (0, helper_1.checkIfActionAccountInitialized)(transactionOptions);
            yield TxManager_1.default.execute(Provider.contract.methods.modifyProvider(providerInfo), transactionOptions);
        });
    }
    /**
     * Function for fetching provider info from blockchain
     */
    getInfo() {
        return __awaiter(this, void 0, void 0, function* () {
            const providerInfoParams = yield Provider.contract.methods
                .getProviderInfo(this.providerId)
                .call()
                .then((providerInfo) => (0, helper_1.cleanWeb3Data)(providerInfo));
            return (this.providerInfo = providerInfoParams);
        });
    }
    /**
     * Function for fetching provider authority address from blockchain
     */
    getAuthority() {
        return this.providerId.toString();
    }
    /**
     * Function for fetching all value offers for this provider
     */
    getValueOffers() {
        return __awaiter(this, void 0, void 0, function* () {
            this.valueOffers = yield Provider.contract.methods
                .getProviderValueOffers(this.providerId)
                .call()
                .then((offers) => offers.map((offer) => offer.toString()));
            return this.valueOffers;
        });
    }
    /**
     * Function for fetching all TEE offers for this provider
     */
    getTeeOffers() {
        return __awaiter(this, void 0, void 0, function* () {
            this.teeOffers = yield Provider.contract.methods
                .getProviderTeeOffers(this.providerId)
                .call()
                .then((offers) => offers.map((offer) => offer.toString()));
            return this.teeOffers;
        });
    }
    /**
     * Function for fetching violationRate for this provider
     */
    getViolationRate() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, helper_1.convertBigIntToString)(yield Provider.contract.methods.getProviderViolationRate(this.providerId).call());
        });
    }
    /**
     * Fetch new Origins (createdDate, createdBy, modifiedDate and modifiedBy)
     */
    getOrigins() {
        return __awaiter(this, void 0, void 0, function* () {
            const origins = yield Provider.contract.methods
                .getProviderOrigins(this.providerId)
                .call()
                .then((origins) => (0, helper_1.cleanWeb3Data)(origins));
            // Convert blockchain time seconds to js time milliseconds
            origins.createdDate = Number(origins.createdDate) * 1000;
            origins.modifiedDate = Number(origins.modifiedDate) * 1000;
            return (this.origins = origins);
        });
    }
    isProviderBanned() {
        return __awaiter(this, void 0, void 0, function* () {
            const violationRate = yield this.getViolationRate();
            const { CONSENSUS_MAX_PENALTIES } = yield Consensus_1.default.getConstants();
            return Number(violationRate) >= Number(CONSENSUS_MAX_PENALTIES);
        });
    }
    getOrdersLockedProfitList() {
        return Provider.contract.methods
            .getOrdersLockedProfitList(this.providerId)
            .call()
            .then((orders) => orders.map((order) => order.toString()));
    }
    getTcbLockedProfitList() {
        return Provider.contract.methods
            .getTcbLockedProfitList(this.providerId)
            .call()
            .then((tcbIds) => tcbIds.map((tcbId) => tcbId.toString()));
    }
}
exports.default = Provider;
