// @super-protocol/sdk-js/build/models/TCB.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const TxManager_1 = __importDefault(require("../utils/TxManager"));
const connectors_1 = require("../connectors");
const helper_1 = require("../utils/helper");
class TCB {
    constructor(tcbId) {
        this.tcbId = tcbId;
        if (!TCB.contract) {
            TCB.contract = connectors_1.BlockchainConnector.getInstance().getContract();
        }
    }
    applyTcbMarks(marks, transactionOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            yield TxManager_1.default.execute(TCB.contract.methods.applyTcbMarks(marks, this.tcbId), transactionOptions);
        });
    }
    setTcbData(pb, quote, transactionOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            (0, helper_1.checkIfActionAccountInitialized)(transactionOptions);
            const fromattedDeviceId = (0, helper_1.packDeviceId)(pb.deviceID);
            yield TxManager_1.default.execute(TCB.contract.methods.setTcbData(this.tcbId, pb.benchmark, pb.properties, fromattedDeviceId, quote), transactionOptions);
        });
    }
    /**
     * Add data to TeeConfirmationBlock and push it to Consensus
     * @param pb - struct of 'processed' data
     * @param quote - data generated from Enclave
     * @param marks - list of marks
     * @param transactionOptions - object what contains alternative action account or gas limit (optional)
     */
    addToSupply(pb, quote, marks, transactionOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            (0, helper_1.checkIfActionAccountInitialized)(transactionOptions);
            yield this.setTcbData(pb, quote, transactionOptions);
            yield this.applyTcbMarks(marks, transactionOptions);
            yield TxManager_1.default.execute(TCB.contract.methods.addTcbToSupply(this.tcbId), transactionOptions);
        });
    }
    /**
     * Assign TCB from SuspiciousBlocks table to check
     * @param transactionOptions - object what contains alternative action account or gas limit (optional)
     */
    assignSuspiciousBlocksToCheck(transactionOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            yield TxManager_1.default.execute(TCB.contract.methods.assignSuspiciousBlocksToCheck(this.tcbId), transactionOptions);
        });
    }
    /**
     * Assign TCB from LastBlocks table to check
     * @param transactionOptions - object what contains alternative action account or gas limit (optional)
     */
    assignLastBlocksToCheck(transactionOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            yield TxManager_1.default.execute(TCB.contract.methods.assignLastBlocksToCheck(this.tcbId), transactionOptions);
        });
    }
    /**
     * Function stake and lock TCB's reward
     * @param tcbId - TEE Offer's completed and valid TCB contract
     * @param transactionOptions - object what contains alternative action account or gas limit (optional)
     */
    claimRewards(transactionOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            (0, helper_1.checkIfActionAccountInitialized)();
            yield TxManager_1.default.execute(TCB.contract.methods.claimRewards(this.tcbId), transactionOptions);
        });
    }
    /**
     * Function unlock previously locked TCB rewards (by claimRewards)
     * @param tcbId - TCB contract address
     * @param unlockAmount - amount of tokens to unlock, max available amount = TeeOffer.getLockInfo(tcbAddress)
     * @param transactionOptions - object what contains alternative action account or gas limit (optional)
     */
    unlockRewards(transactionOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            (0, helper_1.checkIfActionAccountInitialized)();
            yield TxManager_1.default.execute(TCB.contract.methods.unlockTcbReward(this.tcbId), transactionOptions);
        });
    }
    /**
     * Function for fetching all TCB data
     */
    get() {
        return __awaiter(this, void 0, void 0, function* () {
            const tcb = yield TCB.contract.methods.getTcbById(this.tcbId).call();
            tcb.publicData.deviceID = (0, helper_1.unpackDeviceId)(tcb.publicData.deviceID);
            return tcb;
        });
    }
    /**
     * Function for fetching the given marks for recruited TCBs from the Tables of Consensus
     */
    getCheckingBlocksMarks() {
        return __awaiter(this, void 0, void 0, function* () {
            const tcb = yield TCB.contract.methods.getTcbById(this.tcbId).call();
            return {
                blocksIds: tcb.utilData.checkingBlocks,
                marks: tcb.utilData.checkingBlockMarks,
            };
        });
    }
    /**
     * Function for fetching TCB avaliable reward
     */
    getRewardAmount() {
        return TCB.contract.methods.getTcbReward(this.tcbId).call();
    }
}
exports.default = TCB;
